{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "94acffde-a523-4bc6-9fd7-d6c0487fbe02",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/abc.py:106: FutureWarning: xarray subclass CrAnData should explicitly define __slots__\n",
      "  cls = super().__new__(mcls, name, bases, namespace, **kwargs)\n",
      "100%|██████████| 2/2 [00:00<00:00, 14098.50it/s]\n",
      "\u001b[32m2025-03-17 18:46:11.411\u001b[0m | \u001b[1mINFO    \u001b[0m | \u001b[36mcrandata.chrom_io\u001b[0m:\u001b[36mimport_bigwigs\u001b[0m:\u001b[36m417\u001b[0m - \u001b[1mExtracting values from 2 bigWig files...\u001b[0m\n",
      "2it [00:00, 108.68it/s]\n",
      "/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/code/crandata/crandata/crandata.py:61: FutureWarning: Setting attribute 'obs_file_path' on a 'CrAnData' object. Explicitly define __slots__ to suppress this warning for legitimate custom attributes and raise an error when attempting variables assignments.\n",
      "  setattr(self, safe_name, self.data_vars[key])\n",
      "/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/code/crandata/crandata/crandata.py:61: FutureWarning: Setting attribute 'obs_index' on a 'CrAnData' object. Explicitly define __slots__ to suppress this warning for legitimate custom attributes and raise an error when attempting variables assignments.\n",
      "  setattr(self, safe_name, self.data_vars[key])\n",
      "/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/code/crandata/crandata/crandata.py:61: FutureWarning: Setting attribute 'var_chrom' on a 'CrAnData' object. Explicitly define __slots__ to suppress this warning for legitimate custom attributes and raise an error when attempting variables assignments.\n",
      "  setattr(self, safe_name, self.data_vars[key])\n",
      "/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/code/crandata/crandata/crandata.py:61: FutureWarning: Setting attribute 'var_start' on a 'CrAnData' object. Explicitly define __slots__ to suppress this warning for legitimate custom attributes and raise an error when attempting variables assignments.\n",
      "  setattr(self, safe_name, self.data_vars[key])\n",
      "/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/code/crandata/crandata/crandata.py:61: FutureWarning: Setting attribute 'var_end' on a 'CrAnData' object. Explicitly define __slots__ to suppress this warning for legitimate custom attributes and raise an error when attempting variables assignments.\n",
      "  setattr(self, safe_name, self.data_vars[key])\n",
      "/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/code/crandata/crandata/crandata.py:61: FutureWarning: Setting attribute 'var_chunk_index' on a 'CrAnData' object. Explicitly define __slots__ to suppress this warning for legitimate custom attributes and raise an error when attempting variables assignments.\n",
      "  setattr(self, safe_name, self.data_vars[key])\n",
      "/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/code/crandata/crandata/crandata.py:61: FutureWarning: Setting attribute 'var_index' on a 'CrAnData' object. Explicitly define __slots__ to suppress this warning for legitimate custom attributes and raise an error when attempting variables assignments.\n",
      "  setattr(self, safe_name, self.data_vars[key])\n",
      "/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/code/crandata/crandata/crandata.py:61: FutureWarning: Setting attribute 'var_index' on a 'CrAnData' object. Explicitly define __slots__ to suppress this warning for legitimate custom attributes and raise an error when attempting variables assignments.\n",
      "  setattr(self, safe_name, self.data_vars[key])\n",
      "/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/code/crandata/crandata/crandata.py:61: FutureWarning: Setting attribute 'var_chrom' on a 'CrAnData' object. Explicitly define __slots__ to suppress this warning for legitimate custom attributes and raise an error when attempting variables assignments.\n",
      "  setattr(self, safe_name, self.data_vars[key])\n",
      "/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/code/crandata/crandata/crandata.py:61: FutureWarning: Setting attribute 'var_start' on a 'CrAnData' object. Explicitly define __slots__ to suppress this warning for legitimate custom attributes and raise an error when attempting variables assignments.\n",
      "  setattr(self, safe_name, self.data_vars[key])\n",
      "/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/code/crandata/crandata/crandata.py:61: FutureWarning: Setting attribute 'var_end' on a 'CrAnData' object. Explicitly define __slots__ to suppress this warning for legitimate custom attributes and raise an error when attempting variables assignments.\n",
      "  setattr(self, safe_name, self.data_vars[key])\n",
      "/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/code/crandata/crandata/crandata.py:61: FutureWarning: Setting attribute 'var_chunk_index' on a 'CrAnData' object. Explicitly define __slots__ to suppress this warning for legitimate custom attributes and raise an error when attempting variables assignments.\n",
      "  setattr(self, safe_name, self.data_vars[key])\n",
      "/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/code/crandata/crandata/crandata.py:61: FutureWarning: Setting attribute 'obs_file_path' on a 'CrAnData' object. Explicitly define __slots__ to suppress this warning for legitimate custom attributes and raise an error when attempting variables assignments.\n",
      "  setattr(self, safe_name, self.data_vars[key])\n",
      "/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/code/crandata/crandata/crandata.py:61: FutureWarning: Setting attribute 'obs_index' on a 'CrAnData' object. Explicitly define __slots__ to suppress this warning for legitimate custom attributes and raise an error when attempting variables assignments.\n",
      "  setattr(self, safe_name, self.data_vars[key])\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Added Hi-C contact data to adata.varp['hic_contacts']:\n",
      "Shape: (3, 3, 1)\n",
      "<xarray.DataArray (var: 3, var_1: 3, obs: 1)> Size: 140B\n",
      "<COO: shape=(3, 3, 1), dtype=float32, nnz=5, fill_value=0.0>\n",
      "Coordinates:\n",
      "  * var      (var) object 24B 'chr1:100-200' 'chr1:300-400' 'chr1:350-450'\n",
      "  * var_1    (var_1) object 24B 'chr1:100-200' 'chr1:300-400' 'chr1:350-450'\n",
      "  * obs      (obs) <U5 20B 'test2'\n",
      "\n",
      "Directory contents: ['beds', 'bigwigs', 'chrom.sizes', 'consensus.bed', 'chrom_data.nc', 'test2.bedp']\n",
      "\n",
      "Loaded CrAnData from NetCDF:\n",
      "CrAnData object\n",
      "Array names: ['var/index', 'var/chrom', 'var/start', 'var/end', 'var/chunk_index', 'obsm/gex', 'obs/file_path', 'obs/index', 'X', 'always_convert_df', 'varp/hic_contacts']\n",
      "Coordinates: ['var', 'obs', 'seq_bins', 'var_1']\n",
      "\n",
      "<xarray.DataArray 'varp/hic_contacts' (var: 3, var_1: 3, obs: 2)> Size: 448B\n",
      "<COO: shape=(3, 3, 2), dtype=float64, nnz=14, fill_value=0.0>\n",
      "Coordinates:\n",
      "  * var      (var) object 24B 'chr1:100-200' 'chr1:300-400' 'chr1:350-450'\n",
      "  * obs      (obs) object 16B 'test' 'test2'\n",
      "  * var_1    (var_1) object 24B 'chr1:100-200' 'chr1:300-400' 'chr1:350-450'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/code/crandata/crandata/crandata.py:61: FutureWarning: Setting attribute 'encoded_varp_hic_contacts' on a 'CrAnData' object. Explicitly define __slots__ to suppress this warning for legitimate custom attributes and raise an error when attempting variables assignments.\n",
      "  setattr(self, safe_name, self.data_vars[key])\n",
      "/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/code/crandata/crandata/crandata.py:61: FutureWarning: Setting attribute 'var_index' on a 'CrAnData' object. Explicitly define __slots__ to suppress this warning for legitimate custom attributes and raise an error when attempting variables assignments.\n",
      "  setattr(self, safe_name, self.data_vars[key])\n",
      "/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/code/crandata/crandata/crandata.py:61: FutureWarning: Setting attribute 'var_chrom' on a 'CrAnData' object. Explicitly define __slots__ to suppress this warning for legitimate custom attributes and raise an error when attempting variables assignments.\n",
      "  setattr(self, safe_name, self.data_vars[key])\n",
      "/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/code/crandata/crandata/crandata.py:61: FutureWarning: Setting attribute 'var_start' on a 'CrAnData' object. Explicitly define __slots__ to suppress this warning for legitimate custom attributes and raise an error when attempting variables assignments.\n",
      "  setattr(self, safe_name, self.data_vars[key])\n",
      "/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/code/crandata/crandata/crandata.py:61: FutureWarning: Setting attribute 'var_end' on a 'CrAnData' object. Explicitly define __slots__ to suppress this warning for legitimate custom attributes and raise an error when attempting variables assignments.\n",
      "  setattr(self, safe_name, self.data_vars[key])\n",
      "/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/code/crandata/crandata/crandata.py:61: FutureWarning: Setting attribute 'var_chunk_index' on a 'CrAnData' object. Explicitly define __slots__ to suppress this warning for legitimate custom attributes and raise an error when attempting variables assignments.\n",
      "  setattr(self, safe_name, self.data_vars[key])\n",
      "/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/code/crandata/crandata/crandata.py:61: FutureWarning: Setting attribute 'obsm_gex' on a 'CrAnData' object. Explicitly define __slots__ to suppress this warning for legitimate custom attributes and raise an error when attempting variables assignments.\n",
      "  setattr(self, safe_name, self.data_vars[key])\n",
      "/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/code/crandata/crandata/crandata.py:61: FutureWarning: Setting attribute 'obs_file_path' on a 'CrAnData' object. Explicitly define __slots__ to suppress this warning for legitimate custom attributes and raise an error when attempting variables assignments.\n",
      "  setattr(self, safe_name, self.data_vars[key])\n",
      "/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/code/crandata/crandata/crandata.py:61: FutureWarning: Setting attribute 'obs_index' on a 'CrAnData' object. Explicitly define __slots__ to suppress this warning for legitimate custom attributes and raise an error when attempting variables assignments.\n",
      "  setattr(self, safe_name, self.data_vars[key])\n",
      "/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/code/crandata/crandata/crandata.py:61: FutureWarning: Setting attribute 'varp_hic_contacts' on a 'CrAnData' object. Explicitly define __slots__ to suppress this warning for legitimate custom attributes and raise an error when attempting variables assignments.\n",
      "  setattr(self, safe_name, self.data_vars[key])\n",
      "/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/code/crandata/crandata/_anndatamodule.py:288: FutureWarning: Setting attribute 'meta_obs_names' on a 'CrAnData' object. Explicitly define __slots__ to suppress this warning for legitimate custom attributes and raise an error when attempting variables assignments.\n",
      "  adata.meta_obs_names = self.meta_obs_names\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-03-17T18:46:37.504261-0700 INFO Loading sequences into memory...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 3/3 [00:00<00:00, 16534.71it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-03-17T18:46:37.508615-0700 INFO Loading sequences into memory...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "100%|██████████| 3/3 [00:00<00:00, 19478.19it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-03-17T18:46:37.521407-0700 INFO Loading sequences into memory...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "100%|██████████| 3/3 [00:00<00:00, 19181.27it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-03-17T18:46:37.525206-0700 INFO Loading sequences into memory...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "100%|██████████| 3/3 [00:00<00:00, 19388.15it/s]\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/torch/utils/data/sampler.py:76: UserWarning: `data_source` argument is not used and will be removed in 2.2.0.You may still have custom implementation that utilizes it.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Iterating over a couple of training batches from MetaAnnDataModule:\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 33%|███▎      | 1/3 [00:00<00:00, 22.83it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<xarray.DataArray 'X' (obs: 2, seq_bins: 100)> Size: 800B\n",
      "array([[5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.],\n",
      "       [4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.]],\n",
      "      dtype=float32)\n",
      "Coordinates:\n",
      "    var       <U12 48B 'chr1:300-400'\n",
      "  * obs       (obs) object 16B 'test' 'test2'\n",
      "  * seq_bins  (seq_bins) int32 400B 0 1 2 3 4 5 6 7 ... 92 93 94 95 96 97 98 99\n",
      "getitem y (2, 100)\n",
      "<xarray.DataArray 'varp/hic_contacts' (var_1: 3, obs: 2)> Size: 120B\n",
      "<COO: shape=(3, 2), dtype=float64, nnz=5, fill_value=0.0>\n",
      "Coordinates:\n",
      "    var      <U12 48B 'chr1:300-400'\n",
      "  * obs      (obs) object 16B 'test' 'test2'\n",
      "  * var_1    (var_1) object 24B 'chr1:100-200' 'chr1:300-400' 'chr1:350-450'\n",
      "getitem hic (3, 2)\n",
      "<xarray.DataArray 'X' (obs: 2, seq_bins: 100)> Size: 800B\n",
      "array([[5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.],\n",
      "       [4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.]],\n",
      "      dtype=float32)\n",
      "Coordinates:\n",
      "    var       <U12 48B 'chr1:300-400'\n",
      "  * obs       (obs) object 16B 'test' 'test2'\n",
      "  * seq_bins  (seq_bins) int32 400B 0 1 2 3 4 5 6 7 ... 92 93 94 95 96 97 98 99\n",
      "getitem y (2, 100)\n",
      "<xarray.DataArray 'varp/hic_contacts' (var_1: 3, obs: 2)> Size: 120B\n",
      "<COO: shape=(3, 2), dtype=float64, nnz=5, fill_value=0.0>\n",
      "Coordinates:\n",
      "    var      <U12 48B 'chr1:300-400'\n",
      "  * obs      (obs) object 16B 'test' 'test2'\n",
      "  * var_1    (var_1) object 24B 'chr1:100-200' 'chr1:300-400' 'chr1:350-450'\n",
      "getitem hic (3, 2)\n",
      "<xarray.DataArray 'X' (obs: 2, seq_bins: 100)> Size: 800B\n",
      "array([[5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.],\n",
      "       [4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.]],\n",
      "      dtype=float32)\n",
      "Coordinates:\n",
      "    var       <U12 48B 'chr1:100-200'\n",
      "  * obs       (obs) object 16B 'test' 'test2'\n",
      "  * seq_bins  (seq_bins) int32 400B 0 1 2 3 4 5 6 7 ... 92 93 94 95 96 97 98 99\n",
      "getitem y (2, 100)\n",
      "<xarray.DataArray 'varp/hic_contacts' (var_1: 3, obs: 2)> Size: 96B\n",
      "<COO: shape=(3, 2), dtype=float64, nnz=4, fill_value=0.0>\n",
      "Coordinates:\n",
      "    var      <U12 48B 'chr1:100-200'\n",
      "  * obs      (obs) object 16B 'test' 'test2'\n",
      "  * var_1    (var_1) object 24B 'chr1:100-200' 'chr1:300-400' 'chr1:350-450'\n",
      "getitem hic (3, 2)\n",
      "<xarray.DataArray 'X' (obs: 2, seq_bins: 100)> Size: 800B\n",
      "array([[5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.],\n",
      "       [4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.]],\n",
      "      dtype=float32)\n",
      "Coordinates:\n",
      "    var       <U12 48B 'chr1:350-450'\n",
      "  * obs       (obs) object 16B 'test' 'test2'\n",
      "  * seq_bins  (seq_bins) int32 400B 0 1 2 3 4 5 6 7 ... 92 93 94 95 96 97 98 99\n",
      "getitem y (2, 100)\n",
      "<xarray.DataArray 'varp/hic_contacts' (var_1: 3, obs: 2)> Size: 120B\n",
      "<COO: shape=(3, 2), dtype=float64, nnz=5, fill_value=0.0>\n",
      "Coordinates:\n",
      "    var      <U12 48B 'chr1:350-450'\n",
      "  * obs      (obs) object 16B 'test' 'test2'\n",
      "  * var_1    (var_1) object 24B 'chr1:100-200' 'chr1:300-400' 'chr1:350-450'\n",
      "getitem hic (3, 2)\n",
      "Meta Batch 0:\n",
      "sequence\n",
      "  sequence: shape torch.Size([4, 100, 4])\n",
      "y\n",
      "  y: shape torch.Size([4, 2, 100])\n",
      "hic\n",
      "  hic: shape torch.Size([2, 2])\n",
      "gex\n",
      "  gex: shape torch.Size([4, 2, 50])\n",
      "<xarray.DataArray 'X' (obs: 2, seq_bins: 100)> Size: 800B\n",
      "array([[5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.],\n",
      "       [4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.]],\n",
      "      dtype=float32)\n",
      "Coordinates:\n",
      "    var       <U12 48B 'chr1:350-450'\n",
      "  * obs       (obs) object 16B 'test' 'test2'\n",
      "  * seq_bins  (seq_bins) int32 400B 0 1 2 3 4 5 6 7 ... 92 93 94 95 96 97 98 99\n",
      "getitem y (2, 100)\n",
      "<xarray.DataArray 'varp/hic_contacts' (var_1: 3, obs: 2)> Size: 120B\n",
      "<COO: shape=(3, 2), dtype=float64, nnz=5, fill_value=0.0>\n",
      "Coordinates:\n",
      "    var      <U12 48B 'chr1:350-450'\n",
      "  * obs      (obs) object 16B 'test' 'test2'\n",
      "  * var_1    (var_1) object 24B 'chr1:100-200' 'chr1:300-400' 'chr1:350-450'\n",
      "getitem hic (3, 2)\n",
      "<xarray.DataArray 'X' (obs: 2, seq_bins: 100)> Size: 800B\n",
      "array([[5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.],\n",
      "       [4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.]],\n",
      "      dtype=float32)\n",
      "Coordinates:\n",
      "    var       <U12 48B 'chr1:350-450'\n",
      "  * obs       (obs) object 16B 'test' 'test2'\n",
      "  * seq_bins  (seq_bins) int32 400B 0 1 2 3 4 5 6 7 ... 92 93 94 95 96 97 98 99\n",
      "getitem y (2, 100)\n",
      "<xarray.DataArray 'varp/hic_contacts' (var_1: 3, obs: 2)> Size: 120B\n",
      "<COO: shape=(3, 2), dtype=float64, nnz=5, fill_value=0.0>\n",
      "Coordinates:\n",
      "    var      <U12 48B 'chr1:350-450'\n",
      "  * obs      (obs) object 16B 'test' 'test2'\n",
      "  * var_1    (var_1) object 24B 'chr1:100-200' 'chr1:300-400' 'chr1:350-450'\n",
      "getitem hic (3, 2)\n",
      "<xarray.DataArray 'X' (obs: 2, seq_bins: 100)> Size: 800B\n",
      "array([[5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.],\n",
      "       [4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.]],\n",
      "      dtype=float32)\n",
      "Coordinates:\n",
      "    var       <U12 48B 'chr1:350-450'\n",
      "  * obs       (obs) object 16B 'test' 'test2'\n",
      "  * seq_bins  (seq_bins) int32 400B 0 1 2 3 4 5 6 7 ... 92 93 94 95 96 97 98 99\n",
      "getitem y (2, 100)\n",
      "<xarray.DataArray 'varp/hic_contacts' (var_1: 3, obs: 2)> Size: 120B\n",
      "<COO: shape=(3, 2), dtype=float64, nnz=5, fill_value=0.0>\n",
      "Coordinates:\n",
      "    var      <U12 48B 'chr1:350-450'\n",
      "  * obs      (obs) object 16B 'test' 'test2'\n",
      "  * var_1    (var_1) object 24B 'chr1:100-200' 'chr1:300-400' 'chr1:350-450'\n",
      "getitem hic (3, 2)\n",
      "<xarray.DataArray 'X' (obs: 2, seq_bins: 100)> Size: 800B\n",
      "array([[5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
      "        5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.],\n",
      "       [4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
      "        4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.]],\n",
      "      dtype=float32)\n",
      "Coordinates:\n",
      "    var       <U12 48B 'chr1:350-450'\n",
      "  * obs       (obs) object 16B 'test' 'test2'\n",
      "  * seq_bins  (seq_bins) int32 400B 0 1 2 3 4 5 6 7 ... 92 93 94 95 96 97 98 99\n",
      "getitem y (2, 100)\n",
      "<xarray.DataArray 'varp/hic_contacts' (var_1: 3, obs: 2)> Size: 120B\n",
      "<COO: shape=(3, 2), dtype=float64, nnz=5, fill_value=0.0>\n",
      "Coordinates:\n",
      "    var      <U12 48B 'chr1:350-450'\n",
      "  * obs      (obs) object 16B 'test' 'test2'\n",
      "  * var_1    (var_1) object 24B 'chr1:100-200' 'chr1:300-400' 'chr1:350-450'\n",
      "getitem hic (3, 2)\n",
      "Meta Batch 1:\n",
      "sequence\n",
      "  sequence: shape torch.Size([4, 100, 4])\n",
      "y\n",
      "  y: shape torch.Size([4, 2, 100])\n",
      "hic\n",
      "  hic: shape torch.Size([2, 2])\n",
      "gex\n",
      "  gex: shape torch.Size([4, 2, 50])\n",
      "\n",
      "Final directory contents: ['beds', 'bigwigs', 'chrom.sizes', 'consensus.bed', 'chrom_data.nc', 'test2.bedp', 'chr1.fa', 'chr1.fa.fai']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import tempfile\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pyBigWig\n",
    "import copy\n",
    "import importlib\n",
    "import xarray as xr\n",
    "import tqdm\n",
    "\n",
    "# Reload modules to ensure the latest versions are in use.\n",
    "import crandata\n",
    "crandata = importlib.reload(crandata)\n",
    "import crandata.chrom_io\n",
    "crandata.chrom_io = importlib.reload(crandata.chrom_io)\n",
    "import crandata._anndatamodule\n",
    "crandata._anndatamodule = importlib.reload(crandata._anndatamodule)\n",
    "import crandata._dataloader\n",
    "crandata._dataloader = importlib.reload(crandata._dataloader)\n",
    "\n",
    "from crandata._anndatamodule import MetaAnnDataModule\n",
    "from crandata.chrom_io import import_bigwigs, add_contact_strengths_to_varp\n",
    "\n",
    "# Create temporary directories for beds, bigWigs, etc.\n",
    "temp_dir = tempfile.TemporaryDirectory()\n",
    "base_dir = Path(temp_dir.name)\n",
    "beds_dir = base_dir / \"beds\"\n",
    "bigwigs_dir = base_dir / \"bigwigs\"\n",
    "beds_dir.mkdir(exist_ok=True)\n",
    "bigwigs_dir.mkdir(exist_ok=True)\n",
    "\n",
    "# Create a chromsizes file.\n",
    "chromsizes_file = base_dir / \"chrom.sizes\"\n",
    "with open(chromsizes_file, \"w\") as f:\n",
    "    f.write(\"chr1\\t1000\\n\")\n",
    "\n",
    "# Create two BED files (e.g., for two different classes).\n",
    "bed_data_A = pd.DataFrame({0: [\"chr1\", \"chr1\"],\n",
    "                           1: [100, 300],\n",
    "                           2: [200, 400]})\n",
    "bed_data_B = pd.DataFrame({0: [\"chr1\", \"chr1\"],\n",
    "                           1: [150, 350],\n",
    "                           2: [250, 450]})\n",
    "bed_file_A = beds_dir / \"ClassA.bed\"\n",
    "bed_file_B = beds_dir / \"ClassB.bed\"\n",
    "bed_data_A.to_csv(bed_file_A, sep=\"\\t\", header=False, index=False)\n",
    "bed_data_B.to_csv(bed_file_B, sep=\"\\t\", header=False, index=False)\n",
    "\n",
    "# Create a consensus BED file.\n",
    "consensus = pd.DataFrame({0: [\"chr1\", \"chr1\", \"chr1\"],\n",
    "                          1: [100, 300, 350],\n",
    "                          2: [200, 400, 450]})\n",
    "consensus_file = base_dir / \"consensus.bed\"\n",
    "consensus.to_csv(consensus_file, sep=\"\\t\", header=False, index=False)\n",
    "\n",
    "# Create two bigWig files.\n",
    "bigwig_file1 = bigwigs_dir / \"test.bw\"\n",
    "bw1 = pyBigWig.open(str(bigwig_file1), \"w\")\n",
    "bw1.addHeader([(\"chr1\", 1000)])\n",
    "bw1.addEntries(chroms=[\"chr1\"], starts=[0], ends=[1000], values=[5.0])\n",
    "bw1.close()\n",
    "\n",
    "bigwig_file2 = bigwigs_dir / \"test2.bw\"\n",
    "bw2 = pyBigWig.open(str(bigwig_file2), \"w\")\n",
    "bw2.addHeader([(\"chr1\", 1000)])\n",
    "bw2.addEntries(chroms=[\"chr1\"], starts=[0], ends=[1000], values=[4.0])\n",
    "bw2.close()\n",
    "\n",
    "# Set parameters for extraction.\n",
    "target_region_width = 100\n",
    "# Write the extracted data to a NetCDF file.\n",
    "backed_path = base_dir / \"chrom_data.nc\"\n",
    "\n",
    "# Create the CrAnData object from the bigWig files and consensus regions.\n",
    "adata = crandata.chrom_io.import_bigwigs(\n",
    "    bigwigs_folder=bigwigs_dir,\n",
    "    regions_file=consensus_file,\n",
    "    backed_path=str(backed_path),\n",
    "    target_region_width=target_region_width,\n",
    "    chromsizes_file=chromsizes_file,\n",
    ")\n",
    "\n",
    "# Since the new CrAnData is a thin xarray.Dataset, add observation/variable metadata\n",
    "# # by creating separate data arrays (e.g. under keys \"obs/<col>\" and \"var/<col>\").\n",
    "# # Here we create an obs DataFrame from the filenames and a var DataFrame from consensus regions.\n",
    "# obs_df = pd.DataFrame({\n",
    "#     \"file_path\": sorted([str(f) for f in os.listdir(bigwigs_dir) if f.endswith(\".bw\")])\n",
    "# }, index=[os.path.splitext(f)[0].replace(\".\", \"_\") for f in sorted(os.listdir(bigwigs_dir)) if f.endswith(\".bw\")])\n",
    "# var_df = consensus.copy()\n",
    "# var_df[\"region\"] = var_df[0].astype(str) + \":\" + var_df[1].astype(str) + \"-\" + var_df[2].astype(str)\n",
    "# var_df = var_df.set_index(\"region\")\n",
    "\n",
    "# # Save metadata into the CrAnData object.\n",
    "# obs_df.columns = [str(x) for x in obs_df.columns]\n",
    "# for col in obs_df.columns:\n",
    "#     adata[\"obs/\" + col] = xr.DataArray(obs_df[col].values, dims=[\"obs\"])\n",
    "# adata[\"obs/index\"] = xr.DataArray(obs_df.index.values, dims=[\"obs\"])\n",
    "\n",
    "# var_df.columns = [str(x) for x in var_df.columns]\n",
    "# for col in var_df.columns:\n",
    "#     adata[\"var/\" + col] = xr.DataArray(var_df[col].values, dims=[\"var\"])\n",
    "# adata[\"var/index\"] = xr.DataArray(var_df.index.values, dims=[\"var\"])\n",
    "\n",
    "# Add a random obsm entry.\n",
    "adata[\"obsm/gex\"] = xr.DataArray(np.random.randn(adata.sizes[\"obs\"], 50), dims=[\"obs\", \"genes\"])\n",
    "\n",
    "# Create a synthetic BEDP file for Hi-C contacts.\n",
    "synthetic_bedp = pd.DataFrame({\n",
    "    0: [\"chr1\", \"chr1\"],\n",
    "    1: [100, 300],\n",
    "    2: [200, 400],\n",
    "    3: [\"chr1\", \"chr1\"],\n",
    "    4: [150, 350],\n",
    "    5: [250, 450],\n",
    "    6: [10, 20]\n",
    "})\n",
    "synthetic_bedp_file = base_dir / \"test2.bedp\"\n",
    "synthetic_bedp.to_csv(synthetic_bedp_file, sep=\"\\t\", header=False, index=False)\n",
    "\n",
    "# Add Hi-C contact data to adata.varp.\n",
    "contacts = add_contact_strengths_to_varp(adata, [str(synthetic_bedp_file)], key=\"hic_contacts\")\n",
    "print(\"Added Hi-C contact data to adata.varp['hic_contacts']:\")\n",
    "print(\"Shape:\", contacts.shape)\n",
    "print(contacts)\n",
    "\n",
    "# Write the CrAnData object to a NetCDF file and then reload it.\n",
    "adata = adata.sparse_serialized()\n",
    "adata.to_netcdf(str(backed_path))\n",
    "# Reload using the open_dataset class method.\n",
    "adata_loaded = crandata.CrAnData.open_dataset(str(backed_path))\n",
    "print(\"\\nDirectory contents:\", os.listdir(base_dir))\n",
    "print(\"\\nLoaded CrAnData from NetCDF:\")\n",
    "print(adata_loaded)\n",
    "# print(\"obs:\")\n",
    "# print(adata_loaded.obs)\n",
    "# print(\"var:\")\n",
    "# print(adata_loaded.var)\n",
    "print(adata_loaded[\"varp/hic_contacts\"])\n",
    "\n",
    "# Create two copies (simulate two species/datasets) and assign a 'split' column.\n",
    "adata1 = copy.deepcopy(adata_loaded)\n",
    "adata2 = copy.deepcopy(adata_loaded)\n",
    "# For convenience, add a \"split\" column under the var metadata.\n",
    "adata1[\"var/split\"] = xr.DataArray(np.full(adata1.sizes[\"var\"], \"train\"), dims=[\"var\"])\n",
    "adata2[\"var/split\"] = xr.DataArray(np.full(adata2.sizes[\"var\"], \"train\"), dims=[\"var\"])\n",
    "\n",
    "# Create a dummy FASTA file for the genome.\n",
    "fasta_file = base_dir / \"chr1.fa\"\n",
    "with open(fasta_file, \"w\") as f:\n",
    "    f.write(\">chr1\\n\")\n",
    "    f.write(\"A\" * 1000 + \"\\n\")\n",
    "\n",
    "# Create a Genome object.\n",
    "from crandata._genome import Genome\n",
    "dummy_genome = Genome(str(fasta_file), chrom_sizes=str(chromsizes_file))\n",
    "\n",
    "# Instantiate MetaAnnDataModule with the two datasets and corresponding genomes.\n",
    "meta_module = crandata._anndatamodule.MetaAnnDataModule(\n",
    "    adatas=[adata1, adata2],\n",
    "    genomes=[dummy_genome, dummy_genome],\n",
    "    data_sources={'y': 'X', 'hic': 'varp/hic_contacts', 'gex': 'obsm/gex'},\n",
    "    in_memory=True,\n",
    "    random_reverse_complement=True,\n",
    "    max_stochastic_shift=5,\n",
    "    deterministic_shift=False,\n",
    "    shuffle_obs=True,\n",
    "    shuffle=True,\n",
    "    batch_size=4,    # small batch size for testing\n",
    "    epoch_size=10    # small epoch size for quick testing\n",
    ")\n",
    "\n",
    "# Setup the meta module for the \"fit\" stage (train/val split).\n",
    "meta_module.setup(\"fit\")\n",
    "\n",
    "# Retrieve the training dataloader from the meta module and iterate over a couple batches.\n",
    "meta_train_dl = meta_module.train_dataloader\n",
    "print(\"\\nIterating over a couple of training batches from MetaAnnDataModule:\")\n",
    "for i, batch in enumerate(tqdm.tqdm(meta_train_dl.data)):\n",
    "    print(f\"Meta Batch {i}:\")\n",
    "    for key, tensor in batch.items():\n",
    "        print(key)\n",
    "        print(f\"  {key}: shape {tensor.shape}\")\n",
    "    if i >= 1:\n",
    "        break\n",
    "\n",
    "print(\"\\nFinal directory contents:\", os.listdir(base_dir))\n",
    "temp_dir.cleanup()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b0ee0b77-17c3-4e98-b089-558810e1cebf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Array: var/index\n",
      "<xarray.DataArray 'var/index' (var: 3)> Size: 24B\n",
      "array(['chr1:100-200', 'chr1:300-400', 'chr1:350-450'], dtype=object)\n",
      "Coordinates:\n",
      "  * var      (var) object 24B 'chr1:100-200' 'chr1:300-400' 'chr1:350-450'\n",
      "Array: var/chrom\n",
      "<xarray.DataArray 'var/chrom' (var: 3)> Size: 24B\n",
      "array(['chr1', 'chr1', 'chr1'], dtype=object)\n",
      "Coordinates:\n",
      "  * var      (var) object 24B 'chr1:100-200' 'chr1:300-400' 'chr1:350-450'\n",
      "Array: var/start\n",
      "<xarray.DataArray 'var/start' (var: 3)> Size: 12B\n",
      "array([100, 300, 350], dtype=int32)\n",
      "Coordinates:\n",
      "  * var      (var) object 24B 'chr1:100-200' 'chr1:300-400' 'chr1:350-450'\n",
      "Array: var/end\n",
      "<xarray.DataArray 'var/end' (var: 3)> Size: 12B\n",
      "array([200, 400, 450], dtype=int32)\n",
      "Coordinates:\n",
      "  * var      (var) object 24B 'chr1:100-200' 'chr1:300-400' 'chr1:350-450'\n",
      "Array: var/chunk_index\n",
      "<xarray.DataArray 'var/chunk_index' (var: 3)> Size: 12B\n",
      "array([0, 0, 0], dtype=int32)\n",
      "Coordinates:\n",
      "  * var      (var) object 24B 'chr1:100-200' 'chr1:300-400' 'chr1:350-450'\n",
      "Array: obs/file_path\n",
      "<xarray.DataArray 'obs/file_path' (obs: 2)> Size: 16B\n",
      "array(['/scratch/fast/47123/tmpft9rt1m4/bigwigs/test.bw',\n",
      "       '/scratch/fast/47123/tmpft9rt1m4/bigwigs/test2.bw'], dtype=object)\n",
      "Coordinates:\n",
      "  * obs      (obs) object 16B 'test' 'test2'\n",
      "Array: obs/index\n",
      "<xarray.DataArray 'obs/index' (obs: 2)> Size: 16B\n",
      "array(['test', 'test2'], dtype=object)\n",
      "Coordinates:\n",
      "  * obs      (obs) object 16B 'test' 'test2'\n",
      "Array: X\n",
      "<xarray.DataArray 'X' (obs: 2, var: 3, seq_bins: 100)> Size: 2kB\n",
      "array([[[5., 5., ..., 5., 5.],\n",
      "        [5., 5., ..., 5., 5.],\n",
      "        [5., 5., ..., 5., 5.]],\n",
      "\n",
      "       [[4., 4., ..., 4., 4.],\n",
      "        [4., 4., ..., 4., 4.],\n",
      "        [4., 4., ..., 4., 4.]]], dtype=float32)\n",
      "Coordinates:\n",
      "  * var       (var) object 24B 'chr1:100-200' 'chr1:300-400' 'chr1:350-450'\n",
      "  * obs       (obs) object 16B 'test' 'test2'\n",
      "  * seq_bins  (seq_bins) int32 400B 0 1 2 3 4 5 6 7 ... 92 93 94 95 96 97 98 99\n",
      "Array: always_convert_df\n",
      "<xarray.Variable (item: 0)> Size: 0B\n",
      "array([], dtype=float64)\n",
      "Array: obsm/gex\n",
      "<xarray.DataArray 'obsm/gex' (obs: 2, genes: 50)> Size: 800B\n",
      "array([[-0.39219895,  0.07413377, -0.52155907,  2.53583365,  0.42324893,\n",
      "        -0.80681942,  0.5408155 ,  0.1384743 ,  1.84637635,  0.40718356,\n",
      "        -0.03886653, -0.30629784, -1.64005617,  3.75355385, -0.08458356,\n",
      "         0.22335548, -1.27198175, -0.8212836 , -1.78149017, -0.41825952,\n",
      "        -0.1290098 ,  2.44311967,  2.29783389, -0.19923444,  0.79408947,\n",
      "         0.08746005,  1.40356167,  0.29167781,  0.57534268,  1.55698926,\n",
      "         0.66182119,  0.14467371,  0.77643344,  1.15009597,  0.37846174,\n",
      "        -0.84245079, -0.97102444, -0.47394202,  0.62724496, -0.27503408,\n",
      "         1.48356867,  1.70695135,  1.67410269,  0.05491952, -0.55310205,\n",
      "         1.19901714, -0.09365098,  1.36503287,  0.08113759, -1.21861927],\n",
      "       [-2.49064237, -0.50962226, -0.24109978, -1.33024125, -0.94333955,\n",
      "        -0.5104034 ,  1.2329452 , -0.21794937, -0.61619736, -0.8383999 ,\n",
      "         0.60512463, -0.54662448, -1.01472337,  0.5934899 , -0.57387943,\n",
      "        -0.18510479,  0.12002484, -1.8491886 , -0.55907239, -0.95907956,\n",
      "         0.97708493, -0.51681667,  0.55690227,  0.37910704,  0.37519736,\n",
      "        -0.10043931,  1.21017291, -0.09089191, -0.78081761, -0.99760979,\n",
      "         0.71894379, -0.38222085, -1.52647987,  0.11175775, -0.92034595,\n",
      "         0.69172511,  0.09067409, -1.15261928, -0.61387105, -0.82079437,\n",
      "        -0.79289206, -0.6234773 ,  1.24409409, -0.08321023, -0.18553814,\n",
      "         0.49369011, -0.18306491, -0.18924008, -1.58278908, -0.89675893]])\n",
      "Coordinates:\n",
      "  * obs      (obs) object 16B 'test' 'test2'\n",
      "Dimensions without coordinates: genes\n",
      "Array: encoded_varp/hic_contacts\n",
      "<xarray.DataArray 'encoded_varp/hic_contacts' ()> Size: 1kB\n",
      "array('{\"sparse\": true, \"data\": [NaN, 10.0, NaN, NaN, NaN, NaN, 20.0, NaN, 20.0, NaN, NaN, 20.0, NaN, 20.0], \"coords\": [[0, 0, 0, 0, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2], [0, 0, 1, 2, 0, 1, 1, 2, 2, 0, 1, 1, 2, 2], [0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1]], \"shape\": [3, 3, 2], \"dtype\": \"float32\", \"dims\": [\"var\", \"var_1\", \"obs\"]}',\n",
      "      dtype='<U318')\n"
     ]
    }
   ],
   "source": [
    "for n in adata.array_names:\n",
    "    print('Array:',n)\n",
    "    print(adata[n])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "63db6280-ac82-400f-8fe0-6e39f43fee2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/scratch/fast/47123/ipykernel_3828884/2924251009.py:1: FutureWarning: The return type of `Dataset.dims` will be changed to return a set of dimension names in future, in order to be more consistent with `DataArray.dims`. To access a mapping from dimension names to lengths, please use `Dataset.sizes`.\n",
      "  dict(adata.dims)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'var': 3, 'obs': 2, 'seq_bins': 100, 'item': 0, 'genes': 50, 'var_1': 3}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dict(adata.dims)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "75dd30ae-b85b-40ba-a5c2-626186284fb2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Batch 0:\n",
      "  var/index: shape (1,)\n",
      "  var/chrom: shape (1,)\n",
      "  var/start: shape (1,)\n",
      "  var/end: shape (1,)\n",
      "  var/chunk_index: shape (1,)\n",
      "  obs/file_path: shape (2,)\n",
      "  obs/index: shape (2,)\n",
      "  X: shape (2, 1, 100)\n",
      "  always_convert_df: shape (0,)\n",
      "  obsm/gex: shape (2, 50)\n",
      "  encoded_varp/hic_contacts: shape ()\n",
      "Batch 1:\n",
      "  var/index: shape (1,)\n",
      "  var/chrom: shape (1,)\n",
      "  var/start: shape (1,)\n",
      "  var/end: shape (1,)\n",
      "  var/chunk_index: shape (1,)\n",
      "  obs/file_path: shape (2,)\n",
      "  obs/index: shape (2,)\n",
      "  X: shape (2, 1, 100)\n",
      "  always_convert_df: shape (0,)\n",
      "  obsm/gex: shape (2, 50)\n",
      "  encoded_varp/hic_contacts: shape ()\n",
      "Batch 2:\n",
      "  var/index: shape (1,)\n",
      "  var/chrom: shape (1,)\n",
      "  var/start: shape (1,)\n",
      "  var/end: shape (1,)\n",
      "  var/chunk_index: shape (1,)\n",
      "  obs/file_path: shape (2,)\n",
      "  obs/index: shape (2,)\n",
      "  X: shape (2, 1, 100)\n",
      "  always_convert_df: shape (0,)\n",
      "  obsm/gex: shape (2, 50)\n",
      "  encoded_varp/hic_contacts: shape ()\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/scratch/fast/47123/ipykernel_3828884/90032034.py:2: FutureWarning: The return type of `Dataset.dims` will be changed to return a set of dimension names in future, in order to be more consistent with `DataArray.dims`. To access a mapping from dimension names to lengths, please use `Dataset.sizes`.\n",
      "  dim_dict = dict(adata.dims)\n"
     ]
    }
   ],
   "source": [
    "import xbatcher\n",
    "dim_dict = dict(adata.dims)\n",
    "dim_dict['var'] = 1\n",
    "batch_size = 2  # Each batch will include one \"var\" slice.\n",
    "del dim_dict['item']\n",
    "bgen = xbatcher.BatchGenerator(\n",
    "    ds=adata,\n",
    "    input_dims=dim_dict,\n",
    "    batch_dims={'var':2},\n",
    ")\n",
    "\n",
    "# Test the dataloader by iterating over the batches and printing variable shapes.\n",
    "for i, batch in enumerate(bgen):\n",
    "    print(f\"Batch {i}:\")\n",
    "    for var_name, da in batch.data_vars.items():\n",
    "        print(f\"  {var_name}: shape {da.shape}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "8535c697-0b94-45e2-9fa0-92f7f10dd12e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><svg style=\"position: absolute; width: 0; height: 0; overflow: hidden\">\n",
       "<defs>\n",
       "<symbol id=\"icon-database\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z\"></path>\n",
       "<path d=\"M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "<path d=\"M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "</symbol>\n",
       "<symbol id=\"icon-file-text2\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z\"></path>\n",
       "<path d=\"M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "</symbol>\n",
       "</defs>\n",
       "</svg>\n",
       "<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.\n",
       " *\n",
       " */\n",
       "\n",
       ":root {\n",
       "  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));\n",
       "  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));\n",
       "  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));\n",
       "  --xr-border-color: var(--jp-border-color2, #e0e0e0);\n",
       "  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);\n",
       "  --xr-background-color: var(--jp-layout-color0, white);\n",
       "  --xr-background-color-row-even: var(--jp-layout-color1, white);\n",
       "  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);\n",
       "}\n",
       "\n",
       "html[theme=\"dark\"],\n",
       "html[data-theme=\"dark\"],\n",
       "body[data-theme=\"dark\"],\n",
       "body.vscode-dark {\n",
       "  --xr-font-color0: rgba(255, 255, 255, 1);\n",
       "  --xr-font-color2: rgba(255, 255, 255, 0.54);\n",
       "  --xr-font-color3: rgba(255, 255, 255, 0.38);\n",
       "  --xr-border-color: #1f1f1f;\n",
       "  --xr-disabled-color: #515151;\n",
       "  --xr-background-color: #111111;\n",
       "  --xr-background-color-row-even: #111111;\n",
       "  --xr-background-color-row-odd: #313131;\n",
       "}\n",
       "\n",
       ".xr-wrap {\n",
       "  display: block !important;\n",
       "  min-width: 300px;\n",
       "  max-width: 700px;\n",
       "}\n",
       "\n",
       ".xr-text-repr-fallback {\n",
       "  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-header {\n",
       "  padding-top: 6px;\n",
       "  padding-bottom: 6px;\n",
       "  margin-bottom: 4px;\n",
       "  border-bottom: solid 1px var(--xr-border-color);\n",
       "}\n",
       "\n",
       ".xr-header > div,\n",
       ".xr-header > ul {\n",
       "  display: inline;\n",
       "  margin-top: 0;\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-obj-type,\n",
       ".xr-array-name {\n",
       "  margin-left: 2px;\n",
       "  margin-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-obj-type {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-sections {\n",
       "  padding-left: 0 !important;\n",
       "  display: grid;\n",
       "  grid-template-columns: 150px auto auto 1fr 0 20px 0 20px;\n",
       "}\n",
       "\n",
       ".xr-section-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-section-item input {\n",
       "  display: inline-block;\n",
       "  opacity: 0;\n",
       "  height: 0;\n",
       "}\n",
       "\n",
       ".xr-section-item input + label {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label {\n",
       "  cursor: pointer;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-item input:focus + label {\n",
       "  border: 2px solid var(--xr-font-color0);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label:hover {\n",
       "  color: var(--xr-font-color0);\n",
       "}\n",
       "\n",
       ".xr-section-summary {\n",
       "  grid-column: 1;\n",
       "  color: var(--xr-font-color2);\n",
       "  font-weight: 500;\n",
       "}\n",
       "\n",
       ".xr-section-summary > span {\n",
       "  display: inline-block;\n",
       "  padding-left: 0.5em;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in + label:before {\n",
       "  display: inline-block;\n",
       "  content: \"►\";\n",
       "  font-size: 11px;\n",
       "  width: 15px;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label:before {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label:before {\n",
       "  content: \"▼\";\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label > span {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-summary,\n",
       ".xr-section-inline-details {\n",
       "  padding-top: 4px;\n",
       "  padding-bottom: 4px;\n",
       "}\n",
       "\n",
       ".xr-section-inline-details {\n",
       "  grid-column: 2 / -1;\n",
       "}\n",
       "\n",
       ".xr-section-details {\n",
       "  display: none;\n",
       "  grid-column: 1 / -1;\n",
       "  margin-bottom: 5px;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked ~ .xr-section-details {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-array-wrap {\n",
       "  grid-column: 1 / -1;\n",
       "  display: grid;\n",
       "  grid-template-columns: 20px auto;\n",
       "}\n",
       "\n",
       ".xr-array-wrap > label {\n",
       "  grid-column: 1;\n",
       "  vertical-align: top;\n",
       "}\n",
       "\n",
       ".xr-preview {\n",
       "  color: var(--xr-font-color3);\n",
       "}\n",
       "\n",
       ".xr-array-preview,\n",
       ".xr-array-data {\n",
       "  padding: 0 5px !important;\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-array-data,\n",
       ".xr-array-in:checked ~ .xr-array-preview {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-array-in:checked ~ .xr-array-data,\n",
       ".xr-array-preview {\n",
       "  display: inline-block;\n",
       "}\n",
       "\n",
       ".xr-dim-list {\n",
       "  display: inline-block !important;\n",
       "  list-style: none;\n",
       "  padding: 0 !important;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list li {\n",
       "  display: inline-block;\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list:before {\n",
       "  content: \"(\";\n",
       "}\n",
       "\n",
       ".xr-dim-list:after {\n",
       "  content: \")\";\n",
       "}\n",
       "\n",
       ".xr-dim-list li:not(:last-child):after {\n",
       "  content: \",\";\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-has-index {\n",
       "  font-weight: bold;\n",
       "}\n",
       "\n",
       ".xr-var-list,\n",
       ".xr-var-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-var-item > div,\n",
       ".xr-var-item label,\n",
       ".xr-var-item > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-even);\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-var-item > .xr-var-name:hover span {\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-var-list > li:nth-child(odd) > div,\n",
       ".xr-var-list > li:nth-child(odd) > label,\n",
       ".xr-var-list > li:nth-child(odd) > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-odd);\n",
       "}\n",
       "\n",
       ".xr-var-name {\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-var-dims {\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-var-dtype {\n",
       "  grid-column: 3;\n",
       "  text-align: right;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-preview {\n",
       "  grid-column: 4;\n",
       "}\n",
       "\n",
       ".xr-index-preview {\n",
       "  grid-column: 2 / 5;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-name,\n",
       ".xr-var-dims,\n",
       ".xr-var-dtype,\n",
       ".xr-preview,\n",
       ".xr-attrs dt {\n",
       "  white-space: nowrap;\n",
       "  overflow: hidden;\n",
       "  text-overflow: ellipsis;\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-var-name:hover,\n",
       ".xr-var-dims:hover,\n",
       ".xr-var-dtype:hover,\n",
       ".xr-attrs dt:hover {\n",
       "  overflow: visible;\n",
       "  width: auto;\n",
       "  z-index: 1;\n",
       "}\n",
       "\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  display: none;\n",
       "  background-color: var(--xr-background-color) !important;\n",
       "  padding-bottom: 5px !important;\n",
       "}\n",
       "\n",
       ".xr-var-attrs-in:checked ~ .xr-var-attrs,\n",
       ".xr-var-data-in:checked ~ .xr-var-data,\n",
       ".xr-index-data-in:checked ~ .xr-index-data {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       ".xr-var-data > table {\n",
       "  float: right;\n",
       "}\n",
       "\n",
       ".xr-var-name span,\n",
       ".xr-var-data,\n",
       ".xr-index-name div,\n",
       ".xr-index-data,\n",
       ".xr-attrs {\n",
       "  padding-left: 25px !important;\n",
       "}\n",
       "\n",
       ".xr-attrs,\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  grid-column: 1 / -1;\n",
       "}\n",
       "\n",
       "dl.xr-attrs {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  display: grid;\n",
       "  grid-template-columns: 125px auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt,\n",
       ".xr-attrs dd {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  float: left;\n",
       "  padding-right: 10px;\n",
       "  width: auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt {\n",
       "  font-weight: normal;\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-attrs dt:hover span {\n",
       "  display: inline-block;\n",
       "  background: var(--xr-background-color);\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-attrs dd {\n",
       "  grid-column: 2;\n",
       "  white-space: pre-wrap;\n",
       "  word-break: break-all;\n",
       "}\n",
       "\n",
       ".xr-icon-database,\n",
       ".xr-icon-file-text2,\n",
       ".xr-no-icon {\n",
       "  display: inline-block;\n",
       "  vertical-align: middle;\n",
       "  width: 1em;\n",
       "  height: 1.5em !important;\n",
       "  stroke-width: 0;\n",
       "  stroke: currentColor;\n",
       "  fill: currentColor;\n",
       "}\n",
       "</style><pre class='xr-text-repr-fallback'>&lt;xarray.DataArray &#x27;X&#x27; (obs: 2, var: 1, seq_bins: 100)&gt; Size: 800B\n",
       "array([[[5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5.]],\n",
       "\n",
       "       [[4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4.]]], dtype=float32)\n",
       "Coordinates:\n",
       "  * var       (var) object 8B &#x27;chr1:350-450&#x27;\n",
       "  * obs       (obs) object 16B &#x27;test&#x27; &#x27;test2&#x27;\n",
       "  * seq_bins  (seq_bins) int32 400B 0 1 2 3 4 5 6 7 ... 92 93 94 95 96 97 98 99</pre><div class='xr-wrap' style='display:none'><div class='xr-header'><div class='xr-obj-type'>xarray.DataArray</div><div class='xr-array-name'>'X'</div><ul class='xr-dim-list'><li><span class='xr-has-index'>obs</span>: 2</li><li><span class='xr-has-index'>var</span>: 1</li><li><span class='xr-has-index'>seq_bins</span>: 100</li></ul></div><ul class='xr-sections'><li class='xr-section-item'><div class='xr-array-wrap'><input id='section-234009ce-705a-4eb9-9ad7-31e35f8b3cdd' class='xr-array-in' type='checkbox' checked><label for='section-234009ce-705a-4eb9-9ad7-31e35f8b3cdd' title='Show/hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-array-preview xr-preview'><span>5.0 5.0 5.0 5.0 5.0 5.0 5.0 5.0 ... 4.0 4.0 4.0 4.0 4.0 4.0 4.0 4.0</span></div><div class='xr-array-data'><pre>array([[[5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5.]],\n",
       "\n",
       "       [[4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4.]]], dtype=float32)</pre></div></div></li><li class='xr-section-item'><input id='section-00b01528-1d43-4306-8908-1552984822af' class='xr-section-summary-in' type='checkbox'  checked><label for='section-00b01528-1d43-4306-8908-1552984822af' class='xr-section-summary' >Coordinates: <span>(3)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>var</span></div><div class='xr-var-dims'>(var)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;chr1:350-450&#x27;</div><input id='attrs-e94945f7-3da4-48f4-a0f3-4f563775772c' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-e94945f7-3da4-48f4-a0f3-4f563775772c' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-582c42c0-144e-4414-9368-2b9e8a69cc57' class='xr-var-data-in' type='checkbox'><label for='data-582c42c0-144e-4414-9368-2b9e8a69cc57' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;chr1:350-450&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>obs</span></div><div class='xr-var-dims'>(obs)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;test&#x27; &#x27;test2&#x27;</div><input id='attrs-5a8f339d-18f5-4d65-96fb-5bfee238cf14' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-5a8f339d-18f5-4d65-96fb-5bfee238cf14' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-e0ebe4c9-525f-4a94-b18b-9839ada671a2' class='xr-var-data-in' type='checkbox'><label for='data-e0ebe4c9-525f-4a94-b18b-9839ada671a2' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;test&#x27;, &#x27;test2&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>seq_bins</span></div><div class='xr-var-dims'>(seq_bins)</div><div class='xr-var-dtype'>int32</div><div class='xr-var-preview xr-preview'>0 1 2 3 4 5 6 ... 94 95 96 97 98 99</div><input id='attrs-cc10316d-09b3-458e-b8f0-4df651371ea5' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-cc10316d-09b3-458e-b8f0-4df651371ea5' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-55f066bb-9ee6-43e2-9c5e-31806e5a3160' class='xr-var-data-in' type='checkbox'><label for='data-55f066bb-9ee6-43e2-9c5e-31806e5a3160' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,\n",
       "       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35,\n",
       "       36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53,\n",
       "       54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71,\n",
       "       72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89,\n",
       "       90, 91, 92, 93, 94, 95, 96, 97, 98, 99], dtype=int32)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-40e387dd-e7b6-43bf-a7b9-a34154f594da' class='xr-section-summary-in' type='checkbox'  ><label for='section-40e387dd-e7b6-43bf-a7b9-a34154f594da' class='xr-section-summary' >Indexes: <span>(3)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-index-name'><div>var</div></div><div class='xr-index-preview'>PandasIndex</div><input type='checkbox' disabled/><label></label><input id='index-1014d92a-7bda-43f7-8db9-6076ef95d54e' class='xr-index-data-in' type='checkbox'/><label for='index-1014d92a-7bda-43f7-8db9-6076ef95d54e' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([&#x27;chr1:350-450&#x27;], dtype=&#x27;object&#x27;, name=&#x27;var&#x27;))</pre></div></li><li class='xr-var-item'><div class='xr-index-name'><div>obs</div></div><div class='xr-index-preview'>PandasIndex</div><input type='checkbox' disabled/><label></label><input id='index-f4c45a75-eae7-48c8-bb01-82be332aa904' class='xr-index-data-in' type='checkbox'/><label for='index-f4c45a75-eae7-48c8-bb01-82be332aa904' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([&#x27;test&#x27;, &#x27;test2&#x27;], dtype=&#x27;object&#x27;, name=&#x27;obs&#x27;))</pre></div></li><li class='xr-var-item'><div class='xr-index-name'><div>seq_bins</div></div><div class='xr-index-preview'>PandasIndex</div><input type='checkbox' disabled/><label></label><input id='index-3320f829-410d-4516-ae4d-537b6f8bf5c8' class='xr-index-data-in' type='checkbox'/><label for='index-3320f829-410d-4516-ae4d-537b6f8bf5c8' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,\n",
       "       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35,\n",
       "       36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53,\n",
       "       54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71,\n",
       "       72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89,\n",
       "       90, 91, 92, 93, 94, 95, 96, 97, 98, 99],\n",
       "      dtype=&#x27;int32&#x27;, name=&#x27;seq_bins&#x27;))</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-a8ccc1ea-c697-4ee5-a300-ed043cf64c78' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-a8ccc1ea-c697-4ee5-a300-ed043cf64c78' class='xr-section-summary'  title='Expand/collapse section'>Attributes: <span>(0)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'></dl></div></li></ul></div></div>"
      ],
      "text/plain": [
       "<xarray.DataArray 'X' (obs: 2, var: 1, seq_bins: 100)> Size: 800B\n",
       "array([[[5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5.]],\n",
       "\n",
       "       [[4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4.]]], dtype=float32)\n",
       "Coordinates:\n",
       "  * var       (var) object 8B 'chr1:350-450'\n",
       "  * obs       (obs) object 16B 'test' 'test2'\n",
       "  * seq_bins  (seq_bins) int32 400B 0 1 2 3 4 5 6 7 ... 92 93 94 95 96 97 98 99"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "batch['X']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "3bd8ef77-9555-4e8c-a5d8-ea3790d019bc",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "invalid indexer array, does not have integer dtype: array(['chr1:300-400'], dtype='<U12')",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[103], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43madata\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mX\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m[\u001b[49m\u001b[43m:\u001b[49m\u001b[43m,\u001b[49m\u001b[43mnp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43marray\u001b[49m\u001b[43m(\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mchr1:300-400\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m:\u001b[49m\u001b[43m]\u001b[49m\n",
      "File \u001b[0;32m~/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/xarray/core/dataarray.py:904\u001b[0m, in \u001b[0;36mDataArray.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    901\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_coord(key)\n\u001b[1;32m    902\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    903\u001b[0m     \u001b[38;5;66;03m# xarray-style array indexing\u001b[39;00m\n\u001b[0;32m--> 904\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43misel\u001b[49m\u001b[43m(\u001b[49m\u001b[43mindexers\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_item_key_to_dict\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/xarray/core/dataarray.py:1546\u001b[0m, in \u001b[0;36mDataArray.isel\u001b[0;34m(self, indexers, drop, missing_dims, **indexers_kwargs)\u001b[0m\n\u001b[1;32m   1541\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_from_temp_dataset(ds)\n\u001b[1;32m   1543\u001b[0m \u001b[38;5;66;03m# Much faster algorithm for when all indexers are ints, slices, one-dimensional\u001b[39;00m\n\u001b[1;32m   1544\u001b[0m \u001b[38;5;66;03m# lists, or zero or one-dimensional np.ndarray's\u001b[39;00m\n\u001b[0;32m-> 1546\u001b[0m variable \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_variable\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43misel\u001b[49m\u001b[43m(\u001b[49m\u001b[43mindexers\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmissing_dims\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmissing_dims\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1547\u001b[0m indexes, index_variables \u001b[38;5;241m=\u001b[39m isel_indexes(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mxindexes, indexers)\n\u001b[1;32m   1549\u001b[0m coords \u001b[38;5;241m=\u001b[39m {}\n",
      "File \u001b[0;32m~/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/xarray/core/variable.py:1023\u001b[0m, in \u001b[0;36mVariable.isel\u001b[0;34m(self, indexers, missing_dims, **indexers_kwargs)\u001b[0m\n\u001b[1;32m   1020\u001b[0m indexers \u001b[38;5;241m=\u001b[39m drop_dims_from_indexers(indexers, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdims, missing_dims)\n\u001b[1;32m   1022\u001b[0m key \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mtuple\u001b[39m(indexers\u001b[38;5;241m.\u001b[39mget(dim, \u001b[38;5;28mslice\u001b[39m(\u001b[38;5;28;01mNone\u001b[39;00m)) \u001b[38;5;28;01mfor\u001b[39;00m dim \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdims)\n\u001b[0;32m-> 1023\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m[\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m]\u001b[49m\n",
      "File \u001b[0;32m~/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/xarray/core/variable.py:766\u001b[0m, in \u001b[0;36mVariable.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    753\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__getitem__\u001b[39m(\u001b[38;5;28mself\u001b[39m, key) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m Self:\n\u001b[1;32m    754\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Return a new Variable object whose contents are consistent with\u001b[39;00m\n\u001b[1;32m    755\u001b[0m \u001b[38;5;124;03m    getting the provided key from the underlying data.\u001b[39;00m\n\u001b[1;32m    756\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    764\u001b[0m \u001b[38;5;124;03m    array `x.values` directly.\u001b[39;00m\n\u001b[1;32m    765\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 766\u001b[0m     dims, indexer, new_order \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_broadcast_indexes\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    767\u001b[0m     indexable \u001b[38;5;241m=\u001b[39m as_indexable(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_data)\n\u001b[1;32m    769\u001b[0m     data \u001b[38;5;241m=\u001b[39m indexing\u001b[38;5;241m.\u001b[39mapply_indexer(indexable, indexer)\n",
      "File \u001b[0;32m~/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/xarray/core/variable.py:609\u001b[0m, in \u001b[0;36mVariable._broadcast_indexes\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    605\u001b[0m \u001b[38;5;66;03m# Detect it can be mapped as an outer indexer\u001b[39;00m\n\u001b[1;32m    606\u001b[0m \u001b[38;5;66;03m# If all key is unlabeled, or\u001b[39;00m\n\u001b[1;32m    607\u001b[0m \u001b[38;5;66;03m# key can be mapped as an OuterIndexer.\u001b[39;00m\n\u001b[1;32m    608\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mall\u001b[39m(\u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(k, Variable) \u001b[38;5;28;01mfor\u001b[39;00m k \u001b[38;5;129;01min\u001b[39;00m key):\n\u001b[0;32m--> 609\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_broadcast_indexes_outer\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    611\u001b[0m \u001b[38;5;66;03m# If all key is 1-dimensional and there are no duplicate labels,\u001b[39;00m\n\u001b[1;32m    612\u001b[0m \u001b[38;5;66;03m# key can be mapped as an OuterIndexer.\u001b[39;00m\n\u001b[1;32m    613\u001b[0m dims \u001b[38;5;241m=\u001b[39m []\n",
      "File \u001b[0;32m~/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/xarray/core/variable.py:693\u001b[0m, in \u001b[0;36mVariable._broadcast_indexes_outer\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    690\u001b[0m             (k,) \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mnonzero(k)\n\u001b[1;32m    691\u001b[0m     new_key\u001b[38;5;241m.\u001b[39mappend(k)\n\u001b[0;32m--> 693\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m dims, \u001b[43mOuterIndexer\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mtuple\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mnew_key\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m, \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/xarray/core/indexing.py:437\u001b[0m, in \u001b[0;36mOuterIndexer.__init__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    435\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m is_duck_array(k):\n\u001b[1;32m    436\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m np\u001b[38;5;241m.\u001b[39missubdtype(k\u001b[38;5;241m.\u001b[39mdtype, np\u001b[38;5;241m.\u001b[39minteger):\n\u001b[0;32m--> 437\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\n\u001b[1;32m    438\u001b[0m             \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124minvalid indexer array, does not have integer dtype: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mk\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    439\u001b[0m         )\n\u001b[1;32m    440\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m k\u001b[38;5;241m.\u001b[39mndim \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:  \u001b[38;5;66;03m# type: ignore[union-attr]\u001b[39;00m\n\u001b[1;32m    441\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\n\u001b[1;32m    442\u001b[0m             \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124minvalid indexer array for \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mtype\u001b[39m(\u001b[38;5;28mself\u001b[39m)\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m; must be scalar \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    443\u001b[0m             \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mor have 1 dimension: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mk\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    444\u001b[0m         )\n",
      "\u001b[0;31mTypeError\u001b[0m: invalid indexer array, does not have integer dtype: array(['chr1:300-400'], dtype='<U12')"
     ]
    }
   ],
   "source": [
    "adata['X'][:,np.array(['chr1:300-400']),:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "f17c9026-c76d-4e10-a23a-8d3319fffbe8",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "invalid indexer array, does not have integer dtype: array(['chr1:300-400'], dtype='<U12')",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[99], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mbatch\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mX\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m[\u001b[49m\u001b[43m:\u001b[49m\u001b[43m,\u001b[49m\u001b[43mnp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43marray\u001b[49m\u001b[43m(\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mchr1:300-400\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m:\u001b[49m\u001b[43m]\u001b[49m\n",
      "File \u001b[0;32m~/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/xarray/core/dataarray.py:904\u001b[0m, in \u001b[0;36mDataArray.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    901\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_coord(key)\n\u001b[1;32m    902\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    903\u001b[0m     \u001b[38;5;66;03m# xarray-style array indexing\u001b[39;00m\n\u001b[0;32m--> 904\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43misel\u001b[49m\u001b[43m(\u001b[49m\u001b[43mindexers\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_item_key_to_dict\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/xarray/core/dataarray.py:1546\u001b[0m, in \u001b[0;36mDataArray.isel\u001b[0;34m(self, indexers, drop, missing_dims, **indexers_kwargs)\u001b[0m\n\u001b[1;32m   1541\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_from_temp_dataset(ds)\n\u001b[1;32m   1543\u001b[0m \u001b[38;5;66;03m# Much faster algorithm for when all indexers are ints, slices, one-dimensional\u001b[39;00m\n\u001b[1;32m   1544\u001b[0m \u001b[38;5;66;03m# lists, or zero or one-dimensional np.ndarray's\u001b[39;00m\n\u001b[0;32m-> 1546\u001b[0m variable \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_variable\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43misel\u001b[49m\u001b[43m(\u001b[49m\u001b[43mindexers\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmissing_dims\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmissing_dims\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1547\u001b[0m indexes, index_variables \u001b[38;5;241m=\u001b[39m isel_indexes(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mxindexes, indexers)\n\u001b[1;32m   1549\u001b[0m coords \u001b[38;5;241m=\u001b[39m {}\n",
      "File \u001b[0;32m~/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/xarray/core/variable.py:1023\u001b[0m, in \u001b[0;36mVariable.isel\u001b[0;34m(self, indexers, missing_dims, **indexers_kwargs)\u001b[0m\n\u001b[1;32m   1020\u001b[0m indexers \u001b[38;5;241m=\u001b[39m drop_dims_from_indexers(indexers, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdims, missing_dims)\n\u001b[1;32m   1022\u001b[0m key \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mtuple\u001b[39m(indexers\u001b[38;5;241m.\u001b[39mget(dim, \u001b[38;5;28mslice\u001b[39m(\u001b[38;5;28;01mNone\u001b[39;00m)) \u001b[38;5;28;01mfor\u001b[39;00m dim \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdims)\n\u001b[0;32m-> 1023\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m[\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m]\u001b[49m\n",
      "File \u001b[0;32m~/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/xarray/core/variable.py:766\u001b[0m, in \u001b[0;36mVariable.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    753\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m__getitem__\u001b[39m(\u001b[38;5;28mself\u001b[39m, key) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m Self:\n\u001b[1;32m    754\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Return a new Variable object whose contents are consistent with\u001b[39;00m\n\u001b[1;32m    755\u001b[0m \u001b[38;5;124;03m    getting the provided key from the underlying data.\u001b[39;00m\n\u001b[1;32m    756\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    764\u001b[0m \u001b[38;5;124;03m    array `x.values` directly.\u001b[39;00m\n\u001b[1;32m    765\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 766\u001b[0m     dims, indexer, new_order \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_broadcast_indexes\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    767\u001b[0m     indexable \u001b[38;5;241m=\u001b[39m as_indexable(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_data)\n\u001b[1;32m    769\u001b[0m     data \u001b[38;5;241m=\u001b[39m indexing\u001b[38;5;241m.\u001b[39mapply_indexer(indexable, indexer)\n",
      "File \u001b[0;32m~/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/xarray/core/variable.py:609\u001b[0m, in \u001b[0;36mVariable._broadcast_indexes\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    605\u001b[0m \u001b[38;5;66;03m# Detect it can be mapped as an outer indexer\u001b[39;00m\n\u001b[1;32m    606\u001b[0m \u001b[38;5;66;03m# If all key is unlabeled, or\u001b[39;00m\n\u001b[1;32m    607\u001b[0m \u001b[38;5;66;03m# key can be mapped as an OuterIndexer.\u001b[39;00m\n\u001b[1;32m    608\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mall\u001b[39m(\u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(k, Variable) \u001b[38;5;28;01mfor\u001b[39;00m k \u001b[38;5;129;01min\u001b[39;00m key):\n\u001b[0;32m--> 609\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_broadcast_indexes_outer\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    611\u001b[0m \u001b[38;5;66;03m# If all key is 1-dimensional and there are no duplicate labels,\u001b[39;00m\n\u001b[1;32m    612\u001b[0m \u001b[38;5;66;03m# key can be mapped as an OuterIndexer.\u001b[39;00m\n\u001b[1;32m    613\u001b[0m dims \u001b[38;5;241m=\u001b[39m []\n",
      "File \u001b[0;32m~/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/xarray/core/variable.py:693\u001b[0m, in \u001b[0;36mVariable._broadcast_indexes_outer\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    690\u001b[0m             (k,) \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mnonzero(k)\n\u001b[1;32m    691\u001b[0m     new_key\u001b[38;5;241m.\u001b[39mappend(k)\n\u001b[0;32m--> 693\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m dims, \u001b[43mOuterIndexer\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mtuple\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mnew_key\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m, \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/xarray/core/indexing.py:437\u001b[0m, in \u001b[0;36mOuterIndexer.__init__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    435\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m is_duck_array(k):\n\u001b[1;32m    436\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m np\u001b[38;5;241m.\u001b[39missubdtype(k\u001b[38;5;241m.\u001b[39mdtype, np\u001b[38;5;241m.\u001b[39minteger):\n\u001b[0;32m--> 437\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\n\u001b[1;32m    438\u001b[0m             \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124minvalid indexer array, does not have integer dtype: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mk\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    439\u001b[0m         )\n\u001b[1;32m    440\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m k\u001b[38;5;241m.\u001b[39mndim \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:  \u001b[38;5;66;03m# type: ignore[union-attr]\u001b[39;00m\n\u001b[1;32m    441\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\n\u001b[1;32m    442\u001b[0m             \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124minvalid indexer array for \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mtype\u001b[39m(\u001b[38;5;28mself\u001b[39m)\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m; must be scalar \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    443\u001b[0m             \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mor have 1 dimension: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mk\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    444\u001b[0m         )\n",
      "\u001b[0;31mTypeError\u001b[0m: invalid indexer array, does not have integer dtype: array(['chr1:300-400'], dtype='<U12')"
     ]
    }
   ],
   "source": [
    "batch['X'][:,np.array(['chr1:300-400']),:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "75132ee6-7032-4072-97de-c88afd3b1634",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dataset after broadcasting non-'var' variables:\n",
      "<xarray.Dataset> Size: 8kB\n",
      "Dimensions:          (var: 3, obs: 2, seq_bins: 100, genes: 50)\n",
      "Coordinates:\n",
      "  * var              (var) object 24B 'chr1:100-200' ... 'chr1:350-450'\n",
      "Dimensions without coordinates: obs, seq_bins, genes\n",
      "Data variables:\n",
      "    var_index        (var) <U12 144B 'chr1:100-200' ... 'chr1:350-450'\n",
      "    var_chrom        (var) <U4 48B 'chr1' 'chr1' 'chr1'\n",
      "    var_start        (var) int32 12B 100 300 350\n",
      "    var_end          (var) int32 12B 200 400 450\n",
      "    var_chunk_index  (var) int32 12B 0 0 0\n",
      "    obs_file_path    (var, obs) <U48 1kB '/scratch/fast/47004/tmpwdluzk9r/big...\n",
      "    obs_index        (var, obs) <U5 120B 'test' 'test2' ... 'test' 'test2'\n",
      "    X                (obs, var, seq_bins) float32 2kB 5.0 5.0 5.0 ... 4.0 4.0\n",
      "    gex              (var, obs, genes) float64 2kB -1.694 -0.2778 ... -0.5753\n",
      "    hic_contacts     (var) <U101 1kB '{\"sparse\": true, \"data\": [NaN, 10.0, Na...\n",
      "\n",
      "Batch 0:\n",
      "  var_index: dims ('var',), shape (2,)\n",
      "  var_chrom: dims ('var',), shape (2,)\n",
      "  var_start: dims ('var',), shape (2,)\n",
      "  var_end: dims ('var',), shape (2,)\n",
      "  var_chunk_index: dims ('var',), shape (2,)\n",
      "  obs_file_path: dims ('var', 'obs'), shape (2, 2)\n",
      "  obs_index: dims ('var', 'obs'), shape (2, 2)\n",
      "  X: dims ('obs', 'var', 'seq_bins'), shape (2, 2, 100)\n",
      "  gex: dims ('var', 'obs', 'genes'), shape (2, 2, 50)\n",
      "  hic_contacts: dims ('var',), shape (2,)\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/scratch/fast/47004/ipykernel_3815803/2402830859.py:102: FutureWarning: The return type of `Dataset.dims` will be changed to return a set of dimension names in future, in order to be more consistent with `DataArray.dims`. To access a mapping from dimension names to lengths, please use `Dataset.sizes`.\n",
      "  input_dims = dict(ds.dims)\n",
      "/home/matthew.schmitz/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/xbatcher/generators.py:112: UserWarning: The following dimensions were included in both ``input_dims`` and ``batch_dims``. Since ``concat_input_dims`` is ``False``, these dimensions will not impact batch generation: {'var': 2}\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import xarray as xr\n",
    "import numpy as np\n",
    "import xbatcher\n",
    "\n",
    "# ---------------------------\n",
    "# Create the Dataset\n",
    "# ---------------------------\n",
    "# \"var\" arrays (each with dim 'var' of length 3)\n",
    "var_index = xr.DataArray(\n",
    "    ['chr1:100-200', 'chr1:300-400', 'chr1:350-450'],\n",
    "    dims=['var'],\n",
    "    name='var_index'\n",
    ")\n",
    "var_chrom = xr.DataArray(\n",
    "    ['chr1', 'chr1', 'chr1'],\n",
    "    dims=['var'],\n",
    "    name='var_chrom'\n",
    ")\n",
    "var_start = xr.DataArray(\n",
    "    np.array([100, 300, 350], dtype=np.int32),\n",
    "    dims=['var'],\n",
    "    name='var_start'\n",
    ")\n",
    "var_end = xr.DataArray(\n",
    "    np.array([200, 400, 450], dtype=np.int32),\n",
    "    dims=['var'],\n",
    "    name='var_end'\n",
    ")\n",
    "var_chunk_index = xr.DataArray(\n",
    "    np.array([0, 0, 0], dtype=np.int32),\n",
    "    dims=['var'],\n",
    "    name='var_chunk_index'\n",
    ")\n",
    "\n",
    "# \"obs\" arrays (each with dim 'obs' of length 2)\n",
    "obs_file_path = xr.DataArray(\n",
    "    ['/scratch/fast/47004/tmpwdluzk9r/bigwigs/test.bw',\n",
    "     '/scratch/fast/47004/tmpwdluzk9r/bigwigs/test2.bw'],\n",
    "    dims=['obs'],\n",
    "    name='obs_file_path'\n",
    ")\n",
    "obs_index = xr.DataArray(\n",
    "    ['test', 'test2'],\n",
    "    dims=['obs'],\n",
    "    name='obs_index'\n",
    ")\n",
    "\n",
    "# DataArray X with dims ('obs', 'var', 'seq_bins')\n",
    "X = xr.DataArray(\n",
    "    np.full((2, 3, 100), 5, dtype=np.float32),\n",
    "    dims=['obs', 'var', 'seq_bins'],\n",
    "    name='X'\n",
    ")\n",
    "# Change second row values for differentiation\n",
    "X.values[1, :, :] = 4\n",
    "\n",
    "# An additional array defined on obs (or with no 'var')\n",
    "gex = xr.DataArray(\n",
    "    np.random.randn(2, 50),\n",
    "    dims=['obs', 'genes'],\n",
    "    name='gex'\n",
    ")\n",
    "hic_contacts = xr.DataArray(\n",
    "    '{\"sparse\": true, \"data\": [NaN, 10.0, NaN, NaN, NaN, NaN, 20.0, NaN, 20.0, NaN, NaN, 20.0, NaN, 20.0]}',\n",
    "    name='hic_contacts'\n",
    ")\n",
    "\n",
    "# Build the Dataset\n",
    "ds = xr.Dataset({\n",
    "    'var_index': var_index,\n",
    "    'var_chrom': var_chrom,\n",
    "    'var_start': var_start,\n",
    "    'var_end': var_end,\n",
    "    'var_chunk_index': var_chunk_index,\n",
    "    'obs_file_path': obs_file_path,\n",
    "    'obs_index': obs_index,\n",
    "    'X': X,\n",
    "    'gex': gex,\n",
    "    'hic_contacts': hic_contacts,\n",
    "})\n",
    "\n",
    "# ---------------------------\n",
    "# Broadcast variables missing 'var'\n",
    "# ---------------------------\n",
    "# Variables that should have a \"var\" dimension but don't already.\n",
    "non_var_vars = ['obs_file_path', 'obs_index', 'gex', 'hic_contacts']\n",
    "\n",
    "for var in non_var_vars:\n",
    "    if 'var' not in ds[var].dims:\n",
    "        # Expand along 'var' using the coordinate from var_index\n",
    "        ds[var] = ds[var].expand_dims(var=ds['var_index'].values)\n",
    "        \n",
    "print(\"Dataset after broadcasting non-'var' variables:\")\n",
    "print(ds)\n",
    "print()\n",
    "\n",
    "# ---------------------------\n",
    "# Create BatchGenerator slicing along the 'var' dimension\n",
    "# ---------------------------\n",
    "# Specify input_dims to indicate each batch sample should have one \"var\" slice.\n",
    "batch_size = 2\n",
    "input_dims = dict(ds.dims)\n",
    "input_dims['var'] = batch_size\n",
    "bgen = xbatcher.BatchGenerator(\n",
    "    ds=ds,\n",
    "    input_dims=input_dims,\n",
    "    batch_dims={'var': batch_size},\n",
    "    concat_input_dims=False,\n",
    ")\n",
    "\n",
    "# ---------------------------\n",
    "# Test the dataloader by iterating over the batches\n",
    "# ---------------------------\n",
    "for i, batch in enumerate(bgen):\n",
    "    print(f\"Batch {i}:\")\n",
    "    for var_name, da in batch.data_vars.items():\n",
    "        print(f\"  {var_name}: dims {da.dims}, shape {da.shape}\")\n",
    "    print()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "95629916-f8bf-42c6-b49a-83788e7cf096",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><svg style=\"position: absolute; width: 0; height: 0; overflow: hidden\">\n",
       "<defs>\n",
       "<symbol id=\"icon-database\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z\"></path>\n",
       "<path d=\"M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "<path d=\"M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "</symbol>\n",
       "<symbol id=\"icon-file-text2\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z\"></path>\n",
       "<path d=\"M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "</symbol>\n",
       "</defs>\n",
       "</svg>\n",
       "<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.\n",
       " *\n",
       " */\n",
       "\n",
       ":root {\n",
       "  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));\n",
       "  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));\n",
       "  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));\n",
       "  --xr-border-color: var(--jp-border-color2, #e0e0e0);\n",
       "  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);\n",
       "  --xr-background-color: var(--jp-layout-color0, white);\n",
       "  --xr-background-color-row-even: var(--jp-layout-color1, white);\n",
       "  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);\n",
       "}\n",
       "\n",
       "html[theme=\"dark\"],\n",
       "html[data-theme=\"dark\"],\n",
       "body[data-theme=\"dark\"],\n",
       "body.vscode-dark {\n",
       "  --xr-font-color0: rgba(255, 255, 255, 1);\n",
       "  --xr-font-color2: rgba(255, 255, 255, 0.54);\n",
       "  --xr-font-color3: rgba(255, 255, 255, 0.38);\n",
       "  --xr-border-color: #1f1f1f;\n",
       "  --xr-disabled-color: #515151;\n",
       "  --xr-background-color: #111111;\n",
       "  --xr-background-color-row-even: #111111;\n",
       "  --xr-background-color-row-odd: #313131;\n",
       "}\n",
       "\n",
       ".xr-wrap {\n",
       "  display: block !important;\n",
       "  min-width: 300px;\n",
       "  max-width: 700px;\n",
       "}\n",
       "\n",
       ".xr-text-repr-fallback {\n",
       "  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-header {\n",
       "  padding-top: 6px;\n",
       "  padding-bottom: 6px;\n",
       "  margin-bottom: 4px;\n",
       "  border-bottom: solid 1px var(--xr-border-color);\n",
       "}\n",
       "\n",
       ".xr-header > div,\n",
       ".xr-header > ul {\n",
       "  display: inline;\n",
       "  margin-top: 0;\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-obj-type,\n",
       ".xr-array-name {\n",
       "  margin-left: 2px;\n",
       "  margin-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-obj-type {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-sections {\n",
       "  padding-left: 0 !important;\n",
       "  display: grid;\n",
       "  grid-template-columns: 150px auto auto 1fr 0 20px 0 20px;\n",
       "}\n",
       "\n",
       ".xr-section-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-section-item input {\n",
       "  display: inline-block;\n",
       "  opacity: 0;\n",
       "  height: 0;\n",
       "}\n",
       "\n",
       ".xr-section-item input + label {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label {\n",
       "  cursor: pointer;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-item input:focus + label {\n",
       "  border: 2px solid var(--xr-font-color0);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label:hover {\n",
       "  color: var(--xr-font-color0);\n",
       "}\n",
       "\n",
       ".xr-section-summary {\n",
       "  grid-column: 1;\n",
       "  color: var(--xr-font-color2);\n",
       "  font-weight: 500;\n",
       "}\n",
       "\n",
       ".xr-section-summary > span {\n",
       "  display: inline-block;\n",
       "  padding-left: 0.5em;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in + label:before {\n",
       "  display: inline-block;\n",
       "  content: \"►\";\n",
       "  font-size: 11px;\n",
       "  width: 15px;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label:before {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label:before {\n",
       "  content: \"▼\";\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label > span {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-summary,\n",
       ".xr-section-inline-details {\n",
       "  padding-top: 4px;\n",
       "  padding-bottom: 4px;\n",
       "}\n",
       "\n",
       ".xr-section-inline-details {\n",
       "  grid-column: 2 / -1;\n",
       "}\n",
       "\n",
       ".xr-section-details {\n",
       "  display: none;\n",
       "  grid-column: 1 / -1;\n",
       "  margin-bottom: 5px;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked ~ .xr-section-details {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-array-wrap {\n",
       "  grid-column: 1 / -1;\n",
       "  display: grid;\n",
       "  grid-template-columns: 20px auto;\n",
       "}\n",
       "\n",
       ".xr-array-wrap > label {\n",
       "  grid-column: 1;\n",
       "  vertical-align: top;\n",
       "}\n",
       "\n",
       ".xr-preview {\n",
       "  color: var(--xr-font-color3);\n",
       "}\n",
       "\n",
       ".xr-array-preview,\n",
       ".xr-array-data {\n",
       "  padding: 0 5px !important;\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-array-data,\n",
       ".xr-array-in:checked ~ .xr-array-preview {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-array-in:checked ~ .xr-array-data,\n",
       ".xr-array-preview {\n",
       "  display: inline-block;\n",
       "}\n",
       "\n",
       ".xr-dim-list {\n",
       "  display: inline-block !important;\n",
       "  list-style: none;\n",
       "  padding: 0 !important;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list li {\n",
       "  display: inline-block;\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list:before {\n",
       "  content: \"(\";\n",
       "}\n",
       "\n",
       ".xr-dim-list:after {\n",
       "  content: \")\";\n",
       "}\n",
       "\n",
       ".xr-dim-list li:not(:last-child):after {\n",
       "  content: \",\";\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-has-index {\n",
       "  font-weight: bold;\n",
       "}\n",
       "\n",
       ".xr-var-list,\n",
       ".xr-var-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-var-item > div,\n",
       ".xr-var-item label,\n",
       ".xr-var-item > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-even);\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-var-item > .xr-var-name:hover span {\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-var-list > li:nth-child(odd) > div,\n",
       ".xr-var-list > li:nth-child(odd) > label,\n",
       ".xr-var-list > li:nth-child(odd) > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-odd);\n",
       "}\n",
       "\n",
       ".xr-var-name {\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-var-dims {\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-var-dtype {\n",
       "  grid-column: 3;\n",
       "  text-align: right;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-preview {\n",
       "  grid-column: 4;\n",
       "}\n",
       "\n",
       ".xr-index-preview {\n",
       "  grid-column: 2 / 5;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-name,\n",
       ".xr-var-dims,\n",
       ".xr-var-dtype,\n",
       ".xr-preview,\n",
       ".xr-attrs dt {\n",
       "  white-space: nowrap;\n",
       "  overflow: hidden;\n",
       "  text-overflow: ellipsis;\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-var-name:hover,\n",
       ".xr-var-dims:hover,\n",
       ".xr-var-dtype:hover,\n",
       ".xr-attrs dt:hover {\n",
       "  overflow: visible;\n",
       "  width: auto;\n",
       "  z-index: 1;\n",
       "}\n",
       "\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  display: none;\n",
       "  background-color: var(--xr-background-color) !important;\n",
       "  padding-bottom: 5px !important;\n",
       "}\n",
       "\n",
       ".xr-var-attrs-in:checked ~ .xr-var-attrs,\n",
       ".xr-var-data-in:checked ~ .xr-var-data,\n",
       ".xr-index-data-in:checked ~ .xr-index-data {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       ".xr-var-data > table {\n",
       "  float: right;\n",
       "}\n",
       "\n",
       ".xr-var-name span,\n",
       ".xr-var-data,\n",
       ".xr-index-name div,\n",
       ".xr-index-data,\n",
       ".xr-attrs {\n",
       "  padding-left: 25px !important;\n",
       "}\n",
       "\n",
       ".xr-attrs,\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  grid-column: 1 / -1;\n",
       "}\n",
       "\n",
       "dl.xr-attrs {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  display: grid;\n",
       "  grid-template-columns: 125px auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt,\n",
       ".xr-attrs dd {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  float: left;\n",
       "  padding-right: 10px;\n",
       "  width: auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt {\n",
       "  font-weight: normal;\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-attrs dt:hover span {\n",
       "  display: inline-block;\n",
       "  background: var(--xr-background-color);\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-attrs dd {\n",
       "  grid-column: 2;\n",
       "  white-space: pre-wrap;\n",
       "  word-break: break-all;\n",
       "}\n",
       "\n",
       ".xr-icon-database,\n",
       ".xr-icon-file-text2,\n",
       ".xr-no-icon {\n",
       "  display: inline-block;\n",
       "  vertical-align: middle;\n",
       "  width: 1em;\n",
       "  height: 1.5em !important;\n",
       "  stroke-width: 0;\n",
       "  stroke: currentColor;\n",
       "  fill: currentColor;\n",
       "}\n",
       "</style><pre class='xr-text-repr-fallback'>&lt;xarray.DataArray &#x27;X&#x27; (obs: 2, var: 1, seq_bins: 100)&gt; Size: 800B\n",
       "array([[[5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5.]],\n",
       "\n",
       "       [[4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4.]]], dtype=float32)\n",
       "Coordinates:\n",
       "  * var       (var) object 8B &#x27;chr1:350-450&#x27;\n",
       "  * obs       (obs) object 16B &#x27;test&#x27; &#x27;test2&#x27;\n",
       "  * seq_bins  (seq_bins) int32 400B 0 1 2 3 4 5 6 7 ... 92 93 94 95 96 97 98 99</pre><div class='xr-wrap' style='display:none'><div class='xr-header'><div class='xr-obj-type'>xarray.DataArray</div><div class='xr-array-name'>'X'</div><ul class='xr-dim-list'><li><span class='xr-has-index'>obs</span>: 2</li><li><span class='xr-has-index'>var</span>: 1</li><li><span class='xr-has-index'>seq_bins</span>: 100</li></ul></div><ul class='xr-sections'><li class='xr-section-item'><div class='xr-array-wrap'><input id='section-660def60-7372-49ee-ba51-e18cffe857ff' class='xr-array-in' type='checkbox' checked><label for='section-660def60-7372-49ee-ba51-e18cffe857ff' title='Show/hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-array-preview xr-preview'><span>5.0 5.0 5.0 5.0 5.0 5.0 5.0 5.0 ... 4.0 4.0 4.0 4.0 4.0 4.0 4.0 4.0</span></div><div class='xr-array-data'><pre>array([[[5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5.]],\n",
       "\n",
       "       [[4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4.]]], dtype=float32)</pre></div></div></li><li class='xr-section-item'><input id='section-fc2d0be5-2c0c-4c44-9fc6-4d95126364a6' class='xr-section-summary-in' type='checkbox'  checked><label for='section-fc2d0be5-2c0c-4c44-9fc6-4d95126364a6' class='xr-section-summary' >Coordinates: <span>(3)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>var</span></div><div class='xr-var-dims'>(var)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;chr1:350-450&#x27;</div><input id='attrs-9197eb3b-acd8-49d5-b66c-ff241e69aa45' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-9197eb3b-acd8-49d5-b66c-ff241e69aa45' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-b93c0639-c980-4999-8810-94866eedbba2' class='xr-var-data-in' type='checkbox'><label for='data-b93c0639-c980-4999-8810-94866eedbba2' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;chr1:350-450&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>obs</span></div><div class='xr-var-dims'>(obs)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;test&#x27; &#x27;test2&#x27;</div><input id='attrs-6ac5062b-b3ec-4ee7-a39f-1dd4dafb42de' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-6ac5062b-b3ec-4ee7-a39f-1dd4dafb42de' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-be769340-e3c3-4b76-8688-f12d83e0ba05' class='xr-var-data-in' type='checkbox'><label for='data-be769340-e3c3-4b76-8688-f12d83e0ba05' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;test&#x27;, &#x27;test2&#x27;], dtype=object)</pre></div></li><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>seq_bins</span></div><div class='xr-var-dims'>(seq_bins)</div><div class='xr-var-dtype'>int32</div><div class='xr-var-preview xr-preview'>0 1 2 3 4 5 6 ... 94 95 96 97 98 99</div><input id='attrs-a0cce367-f6b7-442e-803d-1ba290f2024c' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-a0cce367-f6b7-442e-803d-1ba290f2024c' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-4a0a777f-662c-4814-87f0-9517845b9993' class='xr-var-data-in' type='checkbox'><label for='data-4a0a777f-662c-4814-87f0-9517845b9993' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,\n",
       "       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35,\n",
       "       36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53,\n",
       "       54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71,\n",
       "       72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89,\n",
       "       90, 91, 92, 93, 94, 95, 96, 97, 98, 99], dtype=int32)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-6176b22f-dffc-44e7-9545-05470a8483a5' class='xr-section-summary-in' type='checkbox'  ><label for='section-6176b22f-dffc-44e7-9545-05470a8483a5' class='xr-section-summary' >Indexes: <span>(3)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-index-name'><div>var</div></div><div class='xr-index-preview'>PandasIndex</div><input type='checkbox' disabled/><label></label><input id='index-221655ad-8574-4c7c-b9c6-f5de0b911976' class='xr-index-data-in' type='checkbox'/><label for='index-221655ad-8574-4c7c-b9c6-f5de0b911976' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([&#x27;chr1:350-450&#x27;], dtype=&#x27;object&#x27;, name=&#x27;var&#x27;))</pre></div></li><li class='xr-var-item'><div class='xr-index-name'><div>obs</div></div><div class='xr-index-preview'>PandasIndex</div><input type='checkbox' disabled/><label></label><input id='index-d8c8cbff-01f1-4b8a-b4be-b6107c669872' class='xr-index-data-in' type='checkbox'/><label for='index-d8c8cbff-01f1-4b8a-b4be-b6107c669872' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([&#x27;test&#x27;, &#x27;test2&#x27;], dtype=&#x27;object&#x27;, name=&#x27;obs&#x27;))</pre></div></li><li class='xr-var-item'><div class='xr-index-name'><div>seq_bins</div></div><div class='xr-index-preview'>PandasIndex</div><input type='checkbox' disabled/><label></label><input id='index-59227f66-28ff-4fbf-b237-99e6174f0553' class='xr-index-data-in' type='checkbox'/><label for='index-59227f66-28ff-4fbf-b237-99e6174f0553' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,\n",
       "       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35,\n",
       "       36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53,\n",
       "       54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71,\n",
       "       72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89,\n",
       "       90, 91, 92, 93, 94, 95, 96, 97, 98, 99],\n",
       "      dtype=&#x27;int32&#x27;, name=&#x27;seq_bins&#x27;))</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-cbc47c3f-6f8b-42c5-910a-a240ed3d5f7d' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-cbc47c3f-6f8b-42c5-910a-a240ed3d5f7d' class='xr-section-summary'  title='Expand/collapse section'>Attributes: <span>(0)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'></dl></div></li></ul></div></div>"
      ],
      "text/plain": [
       "<xarray.DataArray 'X' (obs: 2, var: 1, seq_bins: 100)> Size: 800B\n",
       "array([[[5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.,\n",
       "         5., 5., 5., 5.]],\n",
       "\n",
       "       [[4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,\n",
       "         4., 4., 4., 4.]]], dtype=float32)\n",
       "Coordinates:\n",
       "  * var       (var) object 8B 'chr1:350-450'\n",
       "  * obs       (obs) object 16B 'test' 'test2'\n",
       "  * seq_bins  (seq_bins) int32 400B 0 1 2 3 4 5 6 7 ... 92 93 94 95 96 97 98 99"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "batch['X']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "44defed3-1334-4323-92c6-65ac66cec50f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Coordinates:\n",
       "  * var      (var) object 16B 'chr1:100-200' 'chr1:300-400'"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "batch['X'].coords"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "4f6da6e3-c6e4-4725-bdd9-7dc76e648cfa",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataArray' object has no attribute 'get'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[82], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mbatch\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mX\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m[:,[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mchr1:100-200\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mchr1:300-400\u001b[39m\u001b[38;5;124m'\u001b[39m]]\n",
      "File \u001b[0;32m~/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/xarray/core/common.py:305\u001b[0m, in \u001b[0;36mAttrAccessMixin.__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m    303\u001b[0m         \u001b[38;5;28;01mwith\u001b[39;00m suppress(\u001b[38;5;167;01mKeyError\u001b[39;00m):\n\u001b[1;32m    304\u001b[0m             \u001b[38;5;28;01mreturn\u001b[39;00m source[name]\n\u001b[0;32m--> 305\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mAttributeError\u001b[39;00m(\n\u001b[1;32m    306\u001b[0m     \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mtype\u001b[39m(\u001b[38;5;28mself\u001b[39m)\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m object has no attribute \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mname\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    307\u001b[0m )\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataArray' object has no attribute 'get'"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "d1d6afec-d0ed-48ed-a828-9d96a99bf3d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "FrozenMappingWarningOnValuesAccess({'var': 3, 'obs': 2, 'seq_bins': 100, 'genes': 50, 'item': 0})"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ds.dims"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "2bbe4eb0-b684-419c-8438-33a1cbffb90d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'var': 2, 'obs': 2, 'seq_bins': 100, 'genes': 50, 'item': 0}"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dim_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "31576602-3f68-49ed-8986-8f403b1b92c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0, 1)"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "batch['X'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e687118a-253b-4236-a43c-309a50b0a44d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['var/index',\n",
       " 'var/chrom',\n",
       " 'var/start',\n",
       " 'var/end',\n",
       " 'var/chunk_index',\n",
       " 'obs/file_path',\n",
       " 'obs/index',\n",
       " 'X',\n",
       " 'always_convert_df',\n",
       " 'obsm/gex',\n",
       " 'encoded_varp/hic_contacts']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.array_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ceb2222b-0f4a-47f0-b059-a857aa590ece",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><svg style=\"position: absolute; width: 0; height: 0; overflow: hidden\">\n",
       "<defs>\n",
       "<symbol id=\"icon-database\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M16 0c-8.837 0-16 2.239-16 5v4c0 2.761 7.163 5 16 5s16-2.239 16-5v-4c0-2.761-7.163-5-16-5z\"></path>\n",
       "<path d=\"M16 17c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "<path d=\"M16 26c-8.837 0-16-2.239-16-5v6c0 2.761 7.163 5 16 5s16-2.239 16-5v-6c0 2.761-7.163 5-16 5z\"></path>\n",
       "</symbol>\n",
       "<symbol id=\"icon-file-text2\" viewBox=\"0 0 32 32\">\n",
       "<path d=\"M28.681 7.159c-0.694-0.947-1.662-2.053-2.724-3.116s-2.169-2.030-3.116-2.724c-1.612-1.182-2.393-1.319-2.841-1.319h-15.5c-1.378 0-2.5 1.121-2.5 2.5v27c0 1.378 1.122 2.5 2.5 2.5h23c1.378 0 2.5-1.122 2.5-2.5v-19.5c0-0.448-0.137-1.23-1.319-2.841zM24.543 5.457c0.959 0.959 1.712 1.825 2.268 2.543h-4.811v-4.811c0.718 0.556 1.584 1.309 2.543 2.268zM28 29.5c0 0.271-0.229 0.5-0.5 0.5h-23c-0.271 0-0.5-0.229-0.5-0.5v-27c0-0.271 0.229-0.5 0.5-0.5 0 0 15.499-0 15.5 0v7c0 0.552 0.448 1 1 1h7v19.5z\"></path>\n",
       "<path d=\"M23 26h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 22h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "<path d=\"M23 18h-14c-0.552 0-1-0.448-1-1s0.448-1 1-1h14c0.552 0 1 0.448 1 1s-0.448 1-1 1z\"></path>\n",
       "</symbol>\n",
       "</defs>\n",
       "</svg>\n",
       "<style>/* CSS stylesheet for displaying xarray objects in jupyterlab.\n",
       " *\n",
       " */\n",
       "\n",
       ":root {\n",
       "  --xr-font-color0: var(--jp-content-font-color0, rgba(0, 0, 0, 1));\n",
       "  --xr-font-color2: var(--jp-content-font-color2, rgba(0, 0, 0, 0.54));\n",
       "  --xr-font-color3: var(--jp-content-font-color3, rgba(0, 0, 0, 0.38));\n",
       "  --xr-border-color: var(--jp-border-color2, #e0e0e0);\n",
       "  --xr-disabled-color: var(--jp-layout-color3, #bdbdbd);\n",
       "  --xr-background-color: var(--jp-layout-color0, white);\n",
       "  --xr-background-color-row-even: var(--jp-layout-color1, white);\n",
       "  --xr-background-color-row-odd: var(--jp-layout-color2, #eeeeee);\n",
       "}\n",
       "\n",
       "html[theme=\"dark\"],\n",
       "html[data-theme=\"dark\"],\n",
       "body[data-theme=\"dark\"],\n",
       "body.vscode-dark {\n",
       "  --xr-font-color0: rgba(255, 255, 255, 1);\n",
       "  --xr-font-color2: rgba(255, 255, 255, 0.54);\n",
       "  --xr-font-color3: rgba(255, 255, 255, 0.38);\n",
       "  --xr-border-color: #1f1f1f;\n",
       "  --xr-disabled-color: #515151;\n",
       "  --xr-background-color: #111111;\n",
       "  --xr-background-color-row-even: #111111;\n",
       "  --xr-background-color-row-odd: #313131;\n",
       "}\n",
       "\n",
       ".xr-wrap {\n",
       "  display: block !important;\n",
       "  min-width: 300px;\n",
       "  max-width: 700px;\n",
       "}\n",
       "\n",
       ".xr-text-repr-fallback {\n",
       "  /* fallback to plain text repr when CSS is not injected (untrusted notebook) */\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-header {\n",
       "  padding-top: 6px;\n",
       "  padding-bottom: 6px;\n",
       "  margin-bottom: 4px;\n",
       "  border-bottom: solid 1px var(--xr-border-color);\n",
       "}\n",
       "\n",
       ".xr-header > div,\n",
       ".xr-header > ul {\n",
       "  display: inline;\n",
       "  margin-top: 0;\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-obj-type,\n",
       ".xr-array-name {\n",
       "  margin-left: 2px;\n",
       "  margin-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-obj-type {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-sections {\n",
       "  padding-left: 0 !important;\n",
       "  display: grid;\n",
       "  grid-template-columns: 150px auto auto 1fr 0 20px 0 20px;\n",
       "}\n",
       "\n",
       ".xr-section-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-section-item input {\n",
       "  display: inline-block;\n",
       "  opacity: 0;\n",
       "  height: 0;\n",
       "}\n",
       "\n",
       ".xr-section-item input + label {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label {\n",
       "  cursor: pointer;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-item input:focus + label {\n",
       "  border: 2px solid var(--xr-font-color0);\n",
       "}\n",
       "\n",
       ".xr-section-item input:enabled + label:hover {\n",
       "  color: var(--xr-font-color0);\n",
       "}\n",
       "\n",
       ".xr-section-summary {\n",
       "  grid-column: 1;\n",
       "  color: var(--xr-font-color2);\n",
       "  font-weight: 500;\n",
       "}\n",
       "\n",
       ".xr-section-summary > span {\n",
       "  display: inline-block;\n",
       "  padding-left: 0.5em;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label {\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in + label:before {\n",
       "  display: inline-block;\n",
       "  content: \"►\";\n",
       "  font-size: 11px;\n",
       "  width: 15px;\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:disabled + label:before {\n",
       "  color: var(--xr-disabled-color);\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label:before {\n",
       "  content: \"▼\";\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked + label > span {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-section-summary,\n",
       ".xr-section-inline-details {\n",
       "  padding-top: 4px;\n",
       "  padding-bottom: 4px;\n",
       "}\n",
       "\n",
       ".xr-section-inline-details {\n",
       "  grid-column: 2 / -1;\n",
       "}\n",
       "\n",
       ".xr-section-details {\n",
       "  display: none;\n",
       "  grid-column: 1 / -1;\n",
       "  margin-bottom: 5px;\n",
       "}\n",
       "\n",
       ".xr-section-summary-in:checked ~ .xr-section-details {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-array-wrap {\n",
       "  grid-column: 1 / -1;\n",
       "  display: grid;\n",
       "  grid-template-columns: 20px auto;\n",
       "}\n",
       "\n",
       ".xr-array-wrap > label {\n",
       "  grid-column: 1;\n",
       "  vertical-align: top;\n",
       "}\n",
       "\n",
       ".xr-preview {\n",
       "  color: var(--xr-font-color3);\n",
       "}\n",
       "\n",
       ".xr-array-preview,\n",
       ".xr-array-data {\n",
       "  padding: 0 5px !important;\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-array-data,\n",
       ".xr-array-in:checked ~ .xr-array-preview {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       ".xr-array-in:checked ~ .xr-array-data,\n",
       ".xr-array-preview {\n",
       "  display: inline-block;\n",
       "}\n",
       "\n",
       ".xr-dim-list {\n",
       "  display: inline-block !important;\n",
       "  list-style: none;\n",
       "  padding: 0 !important;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list li {\n",
       "  display: inline-block;\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "}\n",
       "\n",
       ".xr-dim-list:before {\n",
       "  content: \"(\";\n",
       "}\n",
       "\n",
       ".xr-dim-list:after {\n",
       "  content: \")\";\n",
       "}\n",
       "\n",
       ".xr-dim-list li:not(:last-child):after {\n",
       "  content: \",\";\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-has-index {\n",
       "  font-weight: bold;\n",
       "}\n",
       "\n",
       ".xr-var-list,\n",
       ".xr-var-item {\n",
       "  display: contents;\n",
       "}\n",
       "\n",
       ".xr-var-item > div,\n",
       ".xr-var-item label,\n",
       ".xr-var-item > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-even);\n",
       "  margin-bottom: 0;\n",
       "}\n",
       "\n",
       ".xr-var-item > .xr-var-name:hover span {\n",
       "  padding-right: 5px;\n",
       "}\n",
       "\n",
       ".xr-var-list > li:nth-child(odd) > div,\n",
       ".xr-var-list > li:nth-child(odd) > label,\n",
       ".xr-var-list > li:nth-child(odd) > .xr-var-name span {\n",
       "  background-color: var(--xr-background-color-row-odd);\n",
       "}\n",
       "\n",
       ".xr-var-name {\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-var-dims {\n",
       "  grid-column: 2;\n",
       "}\n",
       "\n",
       ".xr-var-dtype {\n",
       "  grid-column: 3;\n",
       "  text-align: right;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-preview {\n",
       "  grid-column: 4;\n",
       "}\n",
       "\n",
       ".xr-index-preview {\n",
       "  grid-column: 2 / 5;\n",
       "  color: var(--xr-font-color2);\n",
       "}\n",
       "\n",
       ".xr-var-name,\n",
       ".xr-var-dims,\n",
       ".xr-var-dtype,\n",
       ".xr-preview,\n",
       ".xr-attrs dt {\n",
       "  white-space: nowrap;\n",
       "  overflow: hidden;\n",
       "  text-overflow: ellipsis;\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-var-name:hover,\n",
       ".xr-var-dims:hover,\n",
       ".xr-var-dtype:hover,\n",
       ".xr-attrs dt:hover {\n",
       "  overflow: visible;\n",
       "  width: auto;\n",
       "  z-index: 1;\n",
       "}\n",
       "\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  display: none;\n",
       "  background-color: var(--xr-background-color) !important;\n",
       "  padding-bottom: 5px !important;\n",
       "}\n",
       "\n",
       ".xr-var-attrs-in:checked ~ .xr-var-attrs,\n",
       ".xr-var-data-in:checked ~ .xr-var-data,\n",
       ".xr-index-data-in:checked ~ .xr-index-data {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       ".xr-var-data > table {\n",
       "  float: right;\n",
       "}\n",
       "\n",
       ".xr-var-name span,\n",
       ".xr-var-data,\n",
       ".xr-index-name div,\n",
       ".xr-index-data,\n",
       ".xr-attrs {\n",
       "  padding-left: 25px !important;\n",
       "}\n",
       "\n",
       ".xr-attrs,\n",
       ".xr-var-attrs,\n",
       ".xr-var-data,\n",
       ".xr-index-data {\n",
       "  grid-column: 1 / -1;\n",
       "}\n",
       "\n",
       "dl.xr-attrs {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  display: grid;\n",
       "  grid-template-columns: 125px auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt,\n",
       ".xr-attrs dd {\n",
       "  padding: 0;\n",
       "  margin: 0;\n",
       "  float: left;\n",
       "  padding-right: 10px;\n",
       "  width: auto;\n",
       "}\n",
       "\n",
       ".xr-attrs dt {\n",
       "  font-weight: normal;\n",
       "  grid-column: 1;\n",
       "}\n",
       "\n",
       ".xr-attrs dt:hover span {\n",
       "  display: inline-block;\n",
       "  background: var(--xr-background-color);\n",
       "  padding-right: 10px;\n",
       "}\n",
       "\n",
       ".xr-attrs dd {\n",
       "  grid-column: 2;\n",
       "  white-space: pre-wrap;\n",
       "  word-break: break-all;\n",
       "}\n",
       "\n",
       ".xr-icon-database,\n",
       ".xr-icon-file-text2,\n",
       ".xr-no-icon {\n",
       "  display: inline-block;\n",
       "  vertical-align: middle;\n",
       "  width: 1em;\n",
       "  height: 1.5em !important;\n",
       "  stroke-width: 0;\n",
       "  stroke: currentColor;\n",
       "  fill: currentColor;\n",
       "}\n",
       "</style><pre class='xr-text-repr-fallback'>&lt;xarray.DataArray &#x27;var/chrom&#x27; (var: 3)&gt; Size: 24B\n",
       "array([&#x27;chr1&#x27;, &#x27;chr1&#x27;, &#x27;chr1&#x27;], dtype=object)\n",
       "Coordinates:\n",
       "  * var      (var) object 24B &#x27;chr1:100-200&#x27; &#x27;chr1:300-400&#x27; &#x27;chr1:350-450&#x27;</pre><div class='xr-wrap' style='display:none'><div class='xr-header'><div class='xr-obj-type'>xarray.DataArray</div><div class='xr-array-name'>'var/chrom'</div><ul class='xr-dim-list'><li><span class='xr-has-index'>var</span>: 3</li></ul></div><ul class='xr-sections'><li class='xr-section-item'><div class='xr-array-wrap'><input id='section-1ade9ea6-dc41-4b3a-b840-2838b1bbdeb5' class='xr-array-in' type='checkbox' checked><label for='section-1ade9ea6-dc41-4b3a-b840-2838b1bbdeb5' title='Show/hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-array-preview xr-preview'><span>&#x27;chr1&#x27; &#x27;chr1&#x27; &#x27;chr1&#x27;</span></div><div class='xr-array-data'><pre>array([&#x27;chr1&#x27;, &#x27;chr1&#x27;, &#x27;chr1&#x27;], dtype=object)</pre></div></div></li><li class='xr-section-item'><input id='section-42648234-61f6-4568-9f39-dbd72eaa49c6' class='xr-section-summary-in' type='checkbox'  checked><label for='section-42648234-61f6-4568-9f39-dbd72eaa49c6' class='xr-section-summary' >Coordinates: <span>(1)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-var-name'><span class='xr-has-index'>var</span></div><div class='xr-var-dims'>(var)</div><div class='xr-var-dtype'>object</div><div class='xr-var-preview xr-preview'>&#x27;chr1:100-200&#x27; ... &#x27;chr1:350-450&#x27;</div><input id='attrs-b1529dfa-6def-4663-a498-9fb0342ccc9b' class='xr-var-attrs-in' type='checkbox' disabled><label for='attrs-b1529dfa-6def-4663-a498-9fb0342ccc9b' title='Show/Hide attributes'><svg class='icon xr-icon-file-text2'><use xlink:href='#icon-file-text2'></use></svg></label><input id='data-7fd3fd37-9af1-4b4d-8374-06f0e70e4b90' class='xr-var-data-in' type='checkbox'><label for='data-7fd3fd37-9af1-4b4d-8374-06f0e70e4b90' title='Show/Hide data repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-var-attrs'><dl class='xr-attrs'></dl></div><div class='xr-var-data'><pre>array([&#x27;chr1:100-200&#x27;, &#x27;chr1:300-400&#x27;, &#x27;chr1:350-450&#x27;], dtype=object)</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-6f262755-15df-47b2-a493-a72840c4f441' class='xr-section-summary-in' type='checkbox'  ><label for='section-6f262755-15df-47b2-a493-a72840c4f441' class='xr-section-summary' >Indexes: <span>(1)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><ul class='xr-var-list'><li class='xr-var-item'><div class='xr-index-name'><div>var</div></div><div class='xr-index-preview'>PandasIndex</div><input type='checkbox' disabled/><label></label><input id='index-8415eebe-7f58-41f9-bee4-ac7446dd4039' class='xr-index-data-in' type='checkbox'/><label for='index-8415eebe-7f58-41f9-bee4-ac7446dd4039' title='Show/Hide index repr'><svg class='icon xr-icon-database'><use xlink:href='#icon-database'></use></svg></label><div class='xr-index-data'><pre>PandasIndex(Index([&#x27;chr1:100-200&#x27;, &#x27;chr1:300-400&#x27;, &#x27;chr1:350-450&#x27;], dtype=&#x27;object&#x27;, name=&#x27;var&#x27;))</pre></div></li></ul></div></li><li class='xr-section-item'><input id='section-d2acdd2c-233c-4498-a44d-36d3d98e7cc7' class='xr-section-summary-in' type='checkbox' disabled ><label for='section-d2acdd2c-233c-4498-a44d-36d3d98e7cc7' class='xr-section-summary'  title='Expand/collapse section'>Attributes: <span>(0)</span></label><div class='xr-section-inline-details'></div><div class='xr-section-details'><dl class='xr-attrs'></dl></div></li></ul></div></div>"
      ],
      "text/plain": [
       "<xarray.DataArray 'var/chrom' (var: 3)> Size: 24B\n",
       "array(['chr1', 'chr1', 'chr1'], dtype=object)\n",
       "Coordinates:\n",
       "  * var      (var) object 24B 'chr1:100-200' 'chr1:300-400' 'chr1:350-450'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata['var/chrom']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c7368b46-88e4-49ee-919b-cdef8fcc2f7b",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "property 'always_convert_df' of 'CrAnData' object has no setter",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[10], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43madata\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43malways_convert_df\u001b[49m \u001b[38;5;241m=\u001b[39m []\n",
      "File \u001b[0;32m~/Matthew/utils/miniforge3/envs/crested/lib/python3.12/site-packages/xarray/core/common.py:316\u001b[0m, in \u001b[0;36mAttrAccessMixin._setattr_dict\u001b[0;34m(self, name, value)\u001b[0m\n\u001b[1;32m    314\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_setattr_dict\u001b[39m(\u001b[38;5;28mself\u001b[39m, name: \u001b[38;5;28mstr\u001b[39m, value: Any) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    315\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Deprecated third party subclass (see ``__init_subclass__`` above)\"\"\"\u001b[39;00m\n\u001b[0;32m--> 316\u001b[0m     \u001b[38;5;28;43mobject\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;21;43m__setattr__\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mname\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mvalue\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    317\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m name \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__dict__\u001b[39m:\n\u001b[1;32m    318\u001b[0m         \u001b[38;5;66;03m# Custom, non-slotted attr, or improperly assigned variable?\u001b[39;00m\n\u001b[1;32m    319\u001b[0m         warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m    320\u001b[0m             \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSetting attribute \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mname\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m on a \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mtype\u001b[39m(\u001b[38;5;28mself\u001b[39m)\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m object. Explicitly define __slots__ \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    321\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mto suppress this warning for legitimate custom attributes and \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    324\u001b[0m             stacklevel\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m2\u001b[39m,\n\u001b[1;32m    325\u001b[0m         )\n",
      "\u001b[0;31mAttributeError\u001b[0m: property 'always_convert_df' of 'CrAnData' object has no setter"
     ]
    }
   ],
   "source": [
    "adata.always_convert_df = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7e8b063-35ce-4e3f-ae52-18512f63bc0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_loaded['obsm/gex']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c39ee3d-4490-45bf-8196-afbcce4dc82a",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_loaded['obs/index']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6601a5e5-1675-48d7-958e-52a8e7db8397",
   "metadata": {},
   "outputs": [],
   "source": [
    "var_df = consensus_peaks.set_index(\"region\")\n",
    "var_df[\"chunk_index\"] = np.arange(var_df.shape[0]) // chunk_size\n",
    "# do something like:\n",
    "var_df = consensus_peaks.set_index(\"region\")\n",
    "var_df[\"chunk_index\"] = np.arange(var_df.shape[0]) // chunk_size\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9b2052d-bccc-4e32-bc54-e957c630746f",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = crandata.CrAnData(always_convert_df=[\"df\"], global_axis_order=[\"obs\"])\n",
    "\n",
    "# Create two 1D arrays of length 10 and store them under hierarchical keys \"df/col1\" and \"df/col2\".\n",
    "adata[\"df/col1\"] = xr.DataArray(np.arange(10), dims=[\"obs\"])\n",
    "adata[\"df/col2\"] = xr.DataArray(np.arange(10, 20), dims=[\"obs\"])\n",
    "\n",
    "# Check that array_names is populated.\n",
    "print(\"Array names:\", adata.array_names)  # Should print: ['df/col1', 'df/col2']\n",
    "\n",
    "# Using get_dataframe() returns a grouped DataFrame.\n",
    "print(\"\\nAccess grouped DataFrame via get_dataframe('df'):\")\n",
    "df_group = adata.get_dataframe(\"df\")\n",
    "print(df_group)\n",
    "\n",
    "# Standard pandas indexing.\n",
    "print(\"\\nAccess a specific element using .loc on the grouped DataFrame:\")\n",
    "print(df_group.loc[5, \"col2\"])\n",
    "\n",
    "# Hierarchical indexing: since \"df\" is in always_convert_df, adata[\"df/col1\"] returns that column.\n",
    "print(\"\\nAccess column 'col1' via hierarchical indexing:\")\n",
    "print(adata[\"df/col1\"])\n",
    "\n",
    "# Dynamic attribute access: \"df/col1\" is accessible as adata.df_col1.\n",
    "print(\"\\nAccess column 'col1' via dynamic attribute (underscore notation):\")\n",
    "print(adata.df_col1)\n",
    "\n",
    "# To see the grouped DataFrame view, call get_dataframe.\n",
    "print(\"\\nAccess grouped DataFrame via get_dataframe('df'):\")\n",
    "print(adata.get_dataframe(\"df\"))\n",
    "\n",
    "# ----- Write to disk and read back -----\n",
    "adata.to_netcdf(\"test_crandata.nc\")\n",
    "print(\"\\nWrote CrAnData to 'test_crandata.nc'.\")\n",
    "\n",
    "# Use our class method to open the file.\n",
    "adata_loaded = crandata.CrAnData.open_dataset(\"test_crandata.nc\")\n",
    "print(\"\\nLoaded CrAnData from disk:\")\n",
    "print(adata_loaded)\n",
    "\n",
    "print(\"\\nAccess grouped DataFrame from loaded data via get_dataframe('df'):\")\n",
    "print(adata_loaded.get_dataframe(\"df\"))\n",
    "\n",
    "print(\"\\nAccess column 'col1' from loaded data using hierarchical indexing:\")\n",
    "print(adata_loaded[\"df/col1\"])\n",
    "\n",
    "os.remove(\"test_crandata.nc\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb929fdb-7478-4f7e-b155-2fa36063fb16",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_loaded.df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84c637bf-5230-4e0d-bf6f-3c9776586345",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import xarray as xr\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import json\n",
    "\n",
    "class CrAnData(xr.Dataset):\n",
    "    __slots__ = (\"always_convert_df\", \"__dict__\")\n",
    "    \n",
    "    def __init__(self, \n",
    "                 data_vars=None,\n",
    "                 coords=None,\n",
    "                 always_convert_df=None,  # list of top-level keys to be grouped into a DataFrame on access\n",
    "                 **kwargs):\n",
    "        \"\"\"\n",
    "        Create a CrAnData object as a subclass of xarray.Dataset.\n",
    "        \n",
    "        Parameters:\n",
    "          data_vars: dictionary of data variables.\n",
    "          coords: dictionary of coordinates.\n",
    "          always_convert_df: list (or array) of top-level keys that should be grouped into a DataFrame on access.\n",
    "          kwargs: additional data variables (merged into data_vars).\n",
    "        \"\"\"\n",
    "        # Merge provided data_vars with kwargs.\n",
    "        if data_vars is None:\n",
    "            data_vars = {}\n",
    "        data_vars = dict(data_vars)  # make a copy\n",
    "        data_vars.update(kwargs)\n",
    "        # Ensure every variable is an xr.DataArray.\n",
    "        for key, var in data_vars.items():\n",
    "            if not isinstance(var, xr.DataArray):\n",
    "                data_vars[key] = xr.DataArray(var)\n",
    "        if coords is None:\n",
    "            coords = {}\n",
    "            \n",
    "        # Initialize the underlying xarray.Dataset.\n",
    "        super().__init__(data_vars=data_vars, coords=coords)\n",
    "        \n",
    "        if always_convert_df is None:\n",
    "            always_convert_df = []\n",
    "        self.always_convert_df = always_convert_df\n",
    "\n",
    "        # Save custom attributes (as JSON) into ds.attrs so they are stored on disk.\n",
    "        self.attrs[\"always_convert_df\"] = json.dumps(self.always_convert_df)\n",
    "        \n",
    "        # For keys that contain '/', add an instance attribute with \"/\" replaced by \"_\" \n",
    "        # for direct dynamic attribute access.\n",
    "        for key in self.data_vars:\n",
    "            if \"/\" in key:\n",
    "                safe_name = key.replace(\"/\", \"_\")\n",
    "                object.__setattr__(self, safe_name, self.data_vars[key])\n",
    "\n",
    "        if \"var\" in self.always_convert_df:\n",
    "            grouped_var = self.get_dataframe(\"var\")\n",
    "            if grouped_var is not None:\n",
    "                self.__dict__[\"var\"] = grouped_var\n",
    "\n",
    "    @property\n",
    "    def array_names(self):\n",
    "        \"\"\"Return a list of the names of the data variables.\"\"\"\n",
    "        return list(self.data_vars.keys())\n",
    "    \n",
    "    def get_dataframe(self, top):\n",
    "        \"\"\"\n",
    "        Group all data variables whose keys start with 'top/' into a pandas DataFrame.\n",
    "        Assumes that each such variable is 1D and they share the same length.\n",
    "        \"\"\"\n",
    "        cols = {}\n",
    "        index = None\n",
    "        for key in list(self.data_vars.keys()):\n",
    "            if key.startswith(top + \"/\"):\n",
    "                col_name = key.split(\"/\", 1)[1]\n",
    "                # Use super().__getitem__ to bypass our custom __getitem__\n",
    "                da = super().__getitem__(key)\n",
    "                cols[col_name] = da.values\n",
    "                if top in da.coords:\n",
    "                    index = da.coords[top].values\n",
    "                else:\n",
    "                    index = np.arange(da.shape[0])\n",
    "        if cols:\n",
    "            return pd.DataFrame(cols, index=index)\n",
    "        else:\n",
    "            return None\n",
    "    \n",
    "    def __getitem__(self, key):\n",
    "        \"\"\"\n",
    "        Support hierarchical indexing.\n",
    "        If key is exactly a top-level key that is in always_convert_df,\n",
    "        return the grouped DataFrame.\n",
    "        If key is a string containing \"/\" and its top-level part is in always_convert_df,\n",
    "        then return the corresponding column from the grouped DataFrame.\n",
    "        Otherwise, return the data variable corresponding to the full key.\n",
    "        \"\"\"\n",
    "        if isinstance(key, str):\n",
    "            if key in self.always_convert_df:\n",
    "                df = self.get_dataframe(key)\n",
    "                if df is None:\n",
    "                    raise KeyError(f\"No grouped data found for key '{key}'\")\n",
    "                return df\n",
    "            if \"/\" in key:\n",
    "                top, sub = key.split(\"/\", 1)\n",
    "                if top in self.always_convert_df:\n",
    "                    df = self.get_dataframe(top)\n",
    "                    if df is None:\n",
    "                        raise KeyError(f\"No grouped data found for key '{top}'\")\n",
    "                    return df[sub]\n",
    "        return super().__getitem__(key)\n",
    "    \n",
    "    def __getattr__(self, attr):\n",
    "        \"\"\"\n",
    "        Fallback attribute lookup.\n",
    "        If attr is one of the keys in always_convert_df (e.g. \"obs\" or \"var\"),\n",
    "        return the grouped DataFrame using get_dataframe.\n",
    "        Otherwise, try to find a data variable whose key (with \"/\" replaced by \"_\") matches attr.\n",
    "        \"\"\"\n",
    "        # If attr is one of the top-level keys to be grouped, return the full DataFrame.\n",
    "        if attr in self.always_convert_df:\n",
    "            df = self.get_dataframe(attr)\n",
    "            if df is not None:\n",
    "                return df\n",
    "        # Otherwise, look for a matching data variable (e.g., \"obs_somecol\" for \"obs/somecol\").\n",
    "        dv = object.__getattribute__(self, \"data_vars\")\n",
    "        for key in dv:\n",
    "            safe = key.replace(\"/\", \"_\")\n",
    "            if safe == attr:\n",
    "                return dv[key]\n",
    "        raise AttributeError(f\"{type(self).__name__!r} object has no attribute {attr!r}\")\n",
    "    \n",
    "    def __repr__(self):\n",
    "        # Custom repr showing our array names, coordinate keys, and global axis order.\n",
    "        rep = f\"CrAnData object\\nArray names: {self.array_names}\\n\"\n",
    "        rep += f\"Coordinates: {list(self.coords.keys())}\\n\"\n",
    "        return rep\n",
    "        \n",
    "    def _repr_html_(self):\n",
    "        return self.__repr__()\n",
    "        \n",
    "    @classmethod\n",
    "    def open_dataset(cls, path, **kwargs):\n",
    "        \"\"\"\n",
    "        Class method to open a NetCDF file and wrap it as a CrAnData object.\n",
    "        Reads custom attributes (always_convert_df) from ds.attrs.\n",
    "        \"\"\"\n",
    "        ds = xr.open_dataset(path, **kwargs)\n",
    "        always_convert_df = json.loads(ds.attrs.get(\"always_convert_df\", \"[]\"))\n",
    "        return cls(data_vars=ds.data_vars, coords=ds.coords,\n",
    "                   always_convert_df=always_convert_df)\n",
    "\n",
    "    @classmethod\n",
    "    def open_zarr(cls, store, **kwargs):\n",
    "        \"\"\"\n",
    "        Class method to open a Zarr store and wrap it as a CrAnData object.\n",
    "        Reads custom attributes (always_convert_df) from ds.attrs.\n",
    "        \"\"\"\n",
    "        ds = xr.open_zarr(store, **kwargs)\n",
    "        always_convert_df = json.loads(ds.attrs.get(\"always_convert_df\", \"[]\"))\n",
    "        return cls(data_vars=ds.data_vars, coords=ds.coords,\n",
    "                   always_convert_df=always_convert_df)\n",
    "\n",
    "# ----- Example usage -----\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    # Create a new CrAnData object.\n",
    "    # Indicate that keys under the top-level \"df\" should be grouped into a DataFrame.\n",
    "    adata = CrAnData(always_convert_df=[\"df\"], global_axis_order=[\"obs\"])\n",
    "    \n",
    "    # Create two 1D arrays of length 10 and store them under hierarchical keys \"df/col1\" and \"df/col2\".\n",
    "    adata[\"df/col1\"] = xr.DataArray(np.arange(10), dims=[\"obs\"])\n",
    "    adata[\"df/col2\"] = xr.DataArray(np.arange(10, 20), dims=[\"obs\"])\n",
    "    \n",
    "    # Check that array_names is populated.\n",
    "    print(\"Array names:\", adata.array_names)  # Should print: ['df/col1', 'df/col2']\n",
    "    \n",
    "    # Using get_dataframe() returns a grouped DataFrame.\n",
    "    print(\"\\nAccess grouped DataFrame via get_dataframe('df'):\")\n",
    "    df_group = adata.get_dataframe(\"df\")\n",
    "    print(df_group)\n",
    "    \n",
    "    # Standard pandas indexing.\n",
    "    print(\"\\nAccess a specific element using .loc on the grouped DataFrame:\")\n",
    "    print(df_group.loc[5, \"col2\"])\n",
    "    \n",
    "    # Hierarchical indexing: since \"df\" is in always_convert_df, adata[\"df/col1\"] returns that column.\n",
    "    print(\"\\nAccess column 'col1' via hierarchical indexing:\")\n",
    "    print(adata[\"df/col1\"])\n",
    "    \n",
    "    # Dynamic attribute access: \"df/col1\" is accessible as adata.df_col1.\n",
    "    print(\"\\nAccess column 'col1' via dynamic attribute (underscore notation):\")\n",
    "    print(adata.df_col1)\n",
    "    \n",
    "    # To see the grouped DataFrame view, call get_dataframe.\n",
    "    print(\"\\nAccess grouped DataFrame via get_dataframe('df'):\")\n",
    "    print(adata.get_dataframe(\"df\"))\n",
    "    \n",
    "    # ----- Write to disk and read back -----\n",
    "    adata.to_netcdf(\"test_crandata.nc\")\n",
    "    print(\"\\nWrote CrAnData to 'test_crandata.nc'.\")\n",
    "    \n",
    "    # Use our class method to open the file.\n",
    "    adata_loaded = CrAnData.open_dataset(\"test_crandata.nc\")\n",
    "    print(\"\\nLoaded CrAnData from disk:\")\n",
    "    print(adata_loaded)\n",
    "\n",
    "    print(\"\\nAccess grouped DataFrame from loaded data via get_dataframe('df'):\")\n",
    "    print(adata_loaded.get_dataframe(\"df\"))\n",
    "    \n",
    "    print(\"\\nAccess column 'col1' from loaded data using hierarchical indexing:\")\n",
    "    print(adata_loaded[\"df/col1\"])\n",
    "    \n",
    "    os.remove(\"test_crandata.nc\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c8480a8-142f-423d-893f-c6074efa3a26",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_loaded.df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0550c266-84c5-49ec-bc4f-fb787c757eb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a2cedbf-3f71-47a3-8361-fbe0511dd321",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af3373ec-309c-4c89-8bb3-81ec43556bbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "It strikes me that this is not a very good way to read from disk, as it might require you to read the whole object into memory and then try to initialize a crandata (does it?). What would it take to create a classmethod for CrAnData that just wrap the xarray open_dataset or open_zarr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be832b9f-8fe5-4fee-bbef-c39648805b16",
   "metadata": {},
   "outputs": [],
   "source": [
    "import crandata\n",
    "import xarray as xr\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9eee12c6-6743-4f06-af25-ebfffd6a9c94",
   "metadata": {},
   "outputs": [],
   "source": [
    "X = xr.DataArray(np.arange(20).reshape(4, 5), dims=[\"obs\", \"var\"])\n",
    "obsm = {\"embedding\": xr.DataArray(np.random.rand(4, 2), dims=[\"obs\", \"other\"])}\n",
    "varm = {\"feature\": xr.DataArray(np.random.rand(5, 3), dims=[\"var\", \"other\"])}\n",
    "layers = {\"layer1\": X.copy()}\n",
    "varp = {\"contacts\": xr.DataArray(np.random.rand(5, 5), dims=[\"var_0\", \"var_1\"])}\n",
    "obsp = {\"adj\": xr.DataArray(np.random.rand(4, 4), dims=[\"obs_0\", \"obs_1\"])}\n",
    "data = crandata.crandata.CrAnData(\n",
    "    X, uns={\"extra\": \"test\"},\n",
    "    obsm=obsm, varm=varm, layers=layers, varp=varp, obsp=obsp\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a825901-0e9e-4fba-8703-1b2b79f66f50",
   "metadata": {},
   "outputs": [],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04edae4a-ab55-4704-a2ec-01e9c2fe9e5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.var['train_probs']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f90c93f-9df3-4724-8d76-bba467e474d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata._propagate_missing_coordinates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7361e12-a505-47a7-9f2e-cb795365b522",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6dde843a-9bd9-4a0d-aa02-5ad3508d2e48",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, batch in enumerate(tqdm.tqdm(meta_train_dl.data)):\n",
    "    print(f\"Meta Batch {i}:\")\n",
    "    for key, tensor in batch.items():\n",
    "        print(f\"  {key}: shape {tensor.shape}\")\n",
    "        # print(tensor)\n",
    "    # For quick testing, you can uncomment the following to break early:\n",
    "    # if i == 1:\n",
    "    #     break\n",
    "\n",
    "print(\"Final directory contents:\", os.listdir(base_dir))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d06f80a2-d874-4823-abaf-58f67726b16e",
   "metadata": {},
   "outputs": [],
   "source": [
    "batch['sequence'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31b4837c-5033-4e05-866f-e4f6a3340bef",
   "metadata": {},
   "outputs": [],
   "source": [
    "batch['hic'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19cf2013-0c20-4563-94d8-bf7c245a47d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.global_axis_order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96f1dd75-1fd4-437d-a720-3e7b2494d423",
   "metadata": {},
   "outputs": [],
   "source": [
    "fff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90ebbd33-d888-4bee-a9b7-93743886288e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import cProfile\n",
    "\n",
    "code = '''\n",
    "for i, batch in enumerate(meta_train_dl.data):\n",
    "    print(f\"Meta Batch {i}:\")\n",
    "    for key, tensor in batch.items():\n",
    "        print(f\"  {key}: shape {tensor.shape}\")\n",
    "'''\n",
    "\n",
    "cProfile.run(code)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c2eb058-5a2f-47e2-9bca-b79d494b1490",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f70292f-f088-4e9d-a127-4736bb591964",
   "metadata": {},
   "outputs": [],
   "source": [
    "import crandata\n",
    "import os\n",
    "import crested\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b454e2de-689f-432e-9621-4d88f73e3b7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "genomes = {}\n",
    "beds = {}\n",
    "chromsizes_files = {}\n",
    "bed_files = {}\n",
    "species = ['mouse','human','macaque']\n",
    "\n",
    "WINDOW_SIZE = 2114\n",
    "OFFSET = WINDOW_SIZE // 2  # e.g., 50% overlap\n",
    "N_THRESHOLD = 0.3\n",
    "n_bins = WINDOW_SIZE//50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "392bdeca-411d-47ea-a846-6c906784b46d",
   "metadata": {},
   "outputs": [],
   "source": [
    "for s in species:\n",
    "    genome_path = '/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/Team/Matthew/genome/onehots/'+s\n",
    "    fasta_file = os.path.join(genome_path,s+'.fa')\n",
    "    chrom_sizes = os.path.join(genome_path,s+'.fa.sizes')\n",
    "    annotation_gtf_file = os.path.join(genome_path,s+'.annotation.gtf')\n",
    "    chromsizes_files[s] = chrom_sizes\n",
    "    genome = crandata.Genome(fasta_file, chrom_sizes, annotation_gtf_file)\n",
    "    genomes[s] = genome\n",
    "    # Set parameters for binning.\n",
    "    \n",
    "    # Optionally specify an output path for the BED file.\n",
    "    OUTPUT_BED = os.path.join(genome_path, \"binned_genome.bed\")\n",
    "    bed_files[s] = OUTPUT_BED\n",
    "    # Generate bins and optionally write to disk.\n",
    "    # binned_df = crandata.bin_genome(genome, WINDOW_SIZE, OFFSET, n_threshold=N_THRESHOLD, output_path=OUTPUT_BED)\n",
    "    # print(\"Filtered bins:\")\n",
    "    # print(binned_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10492d8e-e22d-48a3-8139-6d5aadc02d7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "adatas = {}\n",
    "\n",
    "for s in species:\n",
    "    # bigwigs_dir = os.path.join('/allen/programs/celltypes/workgroups/rnaseqanalysis/EvoGen/SpinalCord/manuscript/ATAC',s,'Group_bigwig')\n",
    "    # adata = crandata.chrom_io.import_bigwigs(\n",
    "    #     bigwigs_folder=bigwigs_dir,\n",
    "    #     regions_file=bed_files[s],\n",
    "    #     backed_path='/home/matthew.schmitz/Matthew/'+s+'_spc_test.h5',\n",
    "    #     target_region_width=WINDOW_SIZE,\n",
    "    #     chromsizes_file=chromsizes_files[s],\n",
    "    #     target = 'mean',\n",
    "    #     n_bins=n_bins\n",
    "    # )\n",
    "    # adatas[s] = adata\n",
    "    adatas[s] = crandata.crandata.CrAnData.from_h5('/home/matthew.schmitz/Matthew/'+s+'_spc_test.h5')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "963ce450-9ef0-4d38-92cc-59521099497c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import numpy as np\n",
    "# adatas['mouse'].uns['chunk_size'] = 512\n",
    "# adatas['human'].uns['chunk_size'] = 512\n",
    "# adatas['macaque'].uns['chunk_size'] = 512\n",
    "# adatas['mouse'].var[\"chunk_index\"] = np.arange(adatas['mouse'].var.shape[0]) // 512\n",
    "# adatas['human'].var[\"chunk_index\"] = np.arange(adatas['human'].var.shape[0]) // 512\n",
    "# adatas['macaque'].var[\"chunk_index\"] = np.arange(adatas['macaque'].var.shape[0]) // 512\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b667671b-3508-4daf-a706-92651c186854",
   "metadata": {},
   "outputs": [],
   "source": [
    "for s in adatas.keys():\n",
    "    crested.pp.train_val_test_split(\n",
    "        adatas[s], strategy=\"region\", val_size=0.1, test_size=0.1, random_state=42\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9aaf99e-b31e-4f93-a15d-e6ca8aa65780",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_module = crandata._anndatamodule.MetaAnnDataModule(\n",
    "    adatas=list(adatas.values()),\n",
    "    genomes=list(genomes.values()),\n",
    "    data_sources={'y': 'X'},\n",
    "    in_memory=False,\n",
    "    random_reverse_complement=True,\n",
    "    max_stochastic_shift=10,\n",
    "    deterministic_shift=False,\n",
    "    shuffle_obs=False, obs_alignment = 'intersect',\n",
    "    shuffle=True,\n",
    "    batch_size=32,    # small batch size for testing\n",
    "    epoch_size=1000000    # small epoch size for quick testing\n",
    ")\n",
    "\n",
    "# Setup the meta module for the \"fit\" stage (train/val)\n",
    "meta_module.setup(\"fit\")\n",
    "\n",
    "# Retrieve the training dataloader from the meta module and iterate over a couple of batches.\n",
    "meta_train_dl = meta_module.train_dataloader\n",
    "\n",
    "print(\"\\nIterating over a couple of training batches from MetaAnnDataModule:\")\n",
    "for i, batch in enumerate(tqdm(meta_train_dl.data)):\n",
    "    print(f\"Meta Batch {i}:\")\n",
    "    for key, tensor in batch.items():\n",
    "        print(f\"  {key}: shape {tensor.shape}\")\n",
    "    if i == 5:\n",
    "        break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de99d0b3-99e8-4a30-8cf4-b762a1247d32",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, batch in enumerate(tqdm(meta_train_dl.data)):\n",
    "    print(f\"Meta Batch {i}:\")\n",
    "    for key, tensor in batch.items():\n",
    "        print(f\"  {key}: shape {tensor.dtype}\")\n",
    "    if i == 5:\n",
    "        break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a11e2cb-b92b-4c02-bc40-886f67d2e8d5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import cProfile\n",
    "\n",
    "code = '''\n",
    "for i, batch in enumerate(meta_train_dl.data):\n",
    "    # print(f\"Meta Batch {i}:\")\n",
    "    # for key, tensor in batch.items():\n",
    "    #     print(f\"  {key}: shape {tensor.shape}\")\n",
    "    if i == 5:\n",
    "        break\n",
    "'''\n",
    "\n",
    "out = cProfile.run(code,sort=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e661681c-6d1e-4441-9878-c21d0505df51",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_architecture = crested.tl.zoo.simple_convnet(\n",
    "    seq_len=2114, num_classes=batch['y'].shape[1]\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6348c92d-cfce-4e39-8a7f-3df53b363b77",
   "metadata": {},
   "outputs": [],
   "source": [
    "import keras\n",
    "# Create your own configuration\n",
    "# I recommend trying this for peak regression with a weighted cosine mse log loss function\n",
    "optimizer = keras.optimizers.Adam(learning_rate=1e-5)\n",
    "loss = crested.tl.losses.CosineMSELogLoss(max_weight=100, multiplier=1)\n",
    "loss = crested.tl.losses.PoissonLoss()\n",
    "\n",
    "metrics = [\n",
    "    keras.metrics.MeanAbsoluteError(),\n",
    "    # keras.metrics.MeanSquaredError(),\n",
    "    # keras.metrics.CosineSimilarity(axis=1),\n",
    "    crested.tl.metrics.PearsonCorrelation(),\n",
    "    # crested.tl.metrics.ConcordanceCorrelationCoefficient(),\n",
    "    # crested.tl.metrics.PearsonCorrelationLog(),\n",
    "    # crested.tl.metrics.ZeroPenaltyMetric(),\n",
    "]\n",
    "\n",
    "alternative_config = crested.tl.TaskConfig(optimizer, loss, metrics)\n",
    "print(alternative_config)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80e5aeb3-5514-466c-b8c7-f04556666c35",
   "metadata": {},
   "outputs": [],
   "source": [
    "# initialize some lazy model parameters *yawn*\n",
    "model_architecture(batch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1b3716a-4a11-418e-9aad-6778019c0f97",
   "metadata": {},
   "outputs": [],
   "source": [
    "trainer = crested.tl.Crested(\n",
    "    data=meta_module,\n",
    "    model=model_architecture,\n",
    "    config=alternative_config,\n",
    "    project_name=\"mouse_biccn\",  # change to your liking\n",
    "    run_name=\"basemodel\",  # change to your liking\n",
    "    logger=None,  # or None, 'dvc', 'tensorboard'\n",
    "    seed=7,  # For reproducibility\n",
    ")\n",
    "# train the model\n",
    "trainer.fit(\n",
    "    epochs=60,\n",
    "    learning_rate_reduce_patience=3,\n",
    "    early_stopping_patience=6,\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9acd5a76-9f07-4525-8c6d-4b5f3ed12aae",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "crested",
   "language": "python",
   "name": "crested"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
